<!doctype html>
<html>
<head>
<meta charset="utf-8" />
<title>The SEEDtk Database</title>
<link href="../css/Basic.css" rel="stylesheet" type="text/css">
</head>
<body>
<div id="Dump">
<h1>The P3-Scripts Interface to PATRIC</h1>

<h2>What is the PATRIC Command-line Interface?</h2>

<a href="https: www.patricbrc.org" target=_blank>PATRIC</a> is an integration of different types of data and software tools
that support research on bacterial pathogens.
The typical biologist seeking access to the PATRIC data and tools will usually explore
the web-based user interface.  However, there are many instances in which programatic or command-line
interfaces are more suitable.  For users that wish command-line access to PATRIC, we provide
the tools described in this document.  We call these tools the <i>P3-scripts</i>.  They
are intended to run on your machine, going over the network to access the services provided by PATRIC.
If you have not already installed a version on your machine, you should probably follow the instructions
discussed in the next section and install a version now.

<h3>Installing the P3-scripts</h3>

<pre>--------------------EVOLVING NOTES -------------------</pre>

<h2>The PATRIC Database</h2>
<p>The main PATRIC database is organized as a series of large, heavily-indexed relational tables. From the
perspective of the CLI, there are five main tables representing objects of interest, connected by four
relationships.</p>
<img src="../img/P3DB.JPG">
<p>The five entities are as follows.</p>
<dl>
<dt>Genome</dt>
<dd>A genome is a set of contigs and annotations representing our best estimate of
the DNA sequence for an organism.
Use <b>p3-all-genomes</b> to list all of the genomes or a subset. Given a list of
genomes,<ul>
<li>Use <b>p3-get-genome-data</b> to retrieve data about the individual genomes.</li>
<li>Use <b>p3-get-genome-features</b> to access the features of the genomes.</li>
<li>Use <b>p3-get-genome-contigs</b> to access the genomes' sequences.</li>
<li>Use <b>p3-get-genome-drugs</b> to access drug resistance data about the genomes.</li>
</ul>
</dd>
<dt>Contig</dt>
<dd>Represents one of the sequences that comprise a genome. A contig can be a chromosome,
a plasmid, or a fragment thereof. Contig data can be accessed from genome IDs using
<b>p3-get-genome-contigs</b>.</dd>
<dt>Drug</dt>
<dd>Represents an antimicrobial drug used for therapeutic treatment. This table is the
anchor for all antimicrobial resistance data in PATRIC. Use <b>p3-all-drugs</b> to get
a list of drugs. Use <b>p3-get-drug-genomes</b> to get resistance data relating to
specific drugs from a list.</dd>
<dt>Feature</dt>
<dd>Represents a region of interest in a genome. This could be a CRISPR array, an
RNA site, a protein encoding region, or a regulatory site, among others. A feature can
be split across multiple regions, or even multiple contigs, but never multiple genomes.
Given a list of genome IDs, use <b>p3-get-genome-features</b> to access the features in
the genomes. Given a list of family IDs, use <b>p3-get-family-features</b> to access the features
in the families. Given a list of feature IDs, use <b>p3-get-feature-data</b> to access data
about those features.</dd>
<dt>Family</dt>
<dd>Represents a protein family, which is a set of features believed to be isofunctional homologs.
Given a list of family IDs, use <b>p3-get-family-features</b>
to get data about the features in the family. Given a list of feature IDs, use <b>p3-get-feature-data</b> to get
the families to which the features belong. There are three types of protein families supported-- <i>local</i>
families which are confined to one genus, <i>global</i> families which cross the entire database, and
<i>figfams</i>, which are computed using a different method.</dd>
</dl>
<h2>Files and Pipelines</h2>

<p>The PATRIC CLI operates on tab-delimited files. That is, each record is divided into fields or columns
separated by tab characters. The first record in each file contains the name of each column. Typically,
a column name consists of a record name, a dot, and a field name. For example, the following file fragment
contains a column from the genome table followed by two columns from the feature table.</p>
<blockquote class="resp">
genome.genome_id    feature.patric_id   feature.product
670.470 fig|670.470.repeat.1    repeat region
670.470 fig|670.470.repeat.2    repeat region
670.470 fig|670.470.repeat.3    repeat region
670.470 fig|670.470.rna.1   tRNA-Ala
670.470 fig|670.470.rna.2   tRNA-Ile
670.470 fig|670.470.repeat.4    repeat region
670.470 fig|670.470.rna.3   16S ribosomal RNA
670.470 fig|670.470.repeat.5    repeat region
670.470 fig|670.470.rna.4   tRNA-Val
670.470 fig|670.470.rna.5   tRNA-Ala
670.470 fig|670.470.repeat.6    repeat region

</blockquote>

<p>The scripts are designed so they can be chained together in pipelines. For example, the above file was generated by the
pipeline</p>
<blockquote class="call">
p3-all-genomes --eq "genome_name,Vibrio parahaemolyticus" | p3-get-genome-features --attr patric_id --attr product

</blockquote>
<p>In this pipeline, the <code>--eq</code> command-line option was used to filter a query, while the
<code>--attr</code> option was used to specify the output columns and the order in which they appear. These
options are available on all of the database scripts.</p>
<p>By default, the last column in the input file is used as the key field for the get-type scripts. You can modify this
behavior using the <code>--col</code> command-line option. The special value <code>0</code> denotes the last column, but
you can also use a 1-based column number (<code>1</code> for the first, <code>2</code> for the second) or a column name.
So, if you want to get location information from the features output by the pipeline above, you could use</p>
<blockquote class="call">
p3-get-feature-data --col=feature.patric_id --attr sequence_id --attr location &lt;input.tbl

</blockquote>
<p>where <i>input.tbl</i> is the above output file. Note that <code>--col=2</code> would also work.</p>
<p>The special script <b>p3-extract</b> allows you to select columns from a file and even change the order. Thus,
the following pipeline removes the genome ID from our file and puts the feature ID at the end before asking
<b>p3-get-feature-data</b> for the location information.</p>
<blockquote class="call">
p3-extract feature.product feature.patric_id &lt;input.tbl | p3-get-feature-data --attr sequence_id --attr location

</blockquote>
<p>Because of the presence of the headings, many standard file-manipulation commands won't work the way you expect. We
provide p3 scripts for several of the most common needs.</p>
<dl>
<dt>p3-extract</dt>
<dd>Select and re-order specific columns.</dd>
<dt>p3-sort</dt>
<dd>Sort by specified columns.</dd>
<dt>p3-match</dt>
<dd>Select records that possess (or do not possess) a particular value in a specified column.</dd>
<dt>p3-pick</dt>
<dd>Randomly select a specified number of records from a file</dd>
</dl>
<h2>Database Script Examples</h2>
<p>In this section we briefly discuss the main database scripts.</p>
<dl>
<dt>p3-all-genomes</dt>
<dd>This script lists all genomes with various characteristics. For example,
<blockquote class="call">
p3-all-genomes --eq genome_name,Streptomyces

</blockquote>
would list all genomes in the genus Streptomyces. (That is, all genomes whose names start with that word.) The <code>--eq</code>
parameter introduces an equality constraint. In PATRIC, string searches perform a substring match, which allows us to easily do
queries of this type. The various database commands all support the <code>--eq</code> option. In addition, you can specify
output fields using the <code>--attr</code> option. Thus,
<blockquote class="call">
p3-all-genomes --eq genome_name,Streptomyces --attr genome_id --attr genome_name

</blockquote>
would output both the ID and name of each genome found, as shown below.
<blockquote class="resp">
genome.genome_id    genome.genome_name
284037.4    Streptomyces sporocinereus strain OsiSh-2
67257.17    Streptomyces albus subsp. albus strain NRRL F-4371
68042.5 Streptomyces hygroscopicus subsp. hygroscopicus strain NBRC 16556
68042.6 Streptomyces hygroscopicus subsp. hygroscopicus strain NBRC 13472
1395572.3   Streptomyces albulus PD-1
68570.5 Streptomyces albulus strain NK660
1160718.3   Streptomyces auratus AGR0001
749414.3    Streptomyces bingchenggensis BCW-1
66876.3 Streptomyces chattanoogensis strain NRRL ISP-5002
249567.6    Streptomyces decoyicus strain NRRL 2666

</blockquote>
To get a complete list of the available fields, use the <code>--fields</code> option. Again, this is available for all the
database scripts described in this section.
<blockquote class="call">
p3-all-genomes --fields

</blockquote>
</dd>
<dt>p3-get-genome-data</dt>
<dd>Given an input file of genome IDs, this script allows you to retrieve additional data and fields. For example, the
following pipeline gets all Streptomyces genomes and then appends the genome name, number of contigs, and DNA length.
<blockquote class="call">
p3-all-genomes --eq genome_name,Streptomyces | p3-get-genome-data --attr genome_name --attr contigs --attr genome_length

</blockquote>
<blockquote class="resp">
genome.genome_id    genome.genome_name  genome.contigs  genome.genome_length
284037.4    Streptomyces sporocinereus strain OsiSh-2   125 10242506
67257.17    Streptomyces albus subsp. albus strain NRRL F-4371  307 9246299
68042.5 Streptomyces hygroscopicus subsp. hygroscopicus strain NBRC 16556   133 10141569
68042.6 Streptomyces hygroscopicus subsp. hygroscopicus strain NBRC 13472   680 9464604
1395572.3   Streptomyces albulus PD-1   425 9340057
68570.5 Streptomyces albulus strain NK660       9372401
1160718.3   Streptomyces auratus AGR0001    213 7825489
749414.3    Streptomyces bingchenggensis BCW-1  0   11936683
66876.3 Streptomyces chattanoogensis strain NRRL ISP-5002   217 9129105

</blockquote>
In actual fact, the use of <b>p3-get-genome-data</b> in the above pipeline is redundant, since <b>p3-all-genomes</b>
supports the same command-line options. In practice, you will use <b>p3-get-genome-data</b> to process genome ID
files created on a separate occasion or via other scripts that don't have the full power for <b>p3-all-genomes</b>.
</dd>
<dt>p3-get-genome-features</dt>
<dd>Given a file of genome IDs, return data about the genome's features. So, for example, the following pipeline would
return the ID and function (product) of each feature in the Mycoplasma genomes.
<blockquote class="call">
p3-all-genomes --eq genome_name,Mycoplasma | p3-get-genome-features --attr patric_id --attr product
</blockquote>
<blockquote class="resp">
genome.genome_id    feature.patric_id   feature.product
1262907.4   fig|1262907.4.repeat.1  repeat region
1262907.4   fig|1262907.4.rna.1 SSU rRNA ## 16S rRNA, small subunit ribosomal RNA
1262907.4   fig|1262907.4.repeat.2  repeat region
1262907.4   fig|1262907.4.peg.1 Cysteinyl-tRNA synthetase related protein
1262907.4   fig|1262907.4.peg.2 hypothetical protein
1262907.4   fig|1262907.4.peg.3 hypothetical protein
1262907.4   fig|1262907.4.peg.4 ABC transporter, ATP-binding protein
1262907.4   fig|1262907.4.peg.5 Glycerol-3-phosphate dehydrogenase [NAD(P)+] (EC 1.1.1.94)
1262907.4   fig|1262907.4.peg.6 BH0638 unknown conserved protein
1262907.4   fig|1262907.4.peg.7 hypothetical protein

</blockquote>
You can use the <code>--fields</code> option to list all the fields available in a feature record. In addition, you have
access to the usual filtering parameters-- <code>--eq</code> as well as <code>--lt</code>, <code>--gt</code>, <code>--le</code>,
<code>--ge</code>, and <code>--ne</code>. So, for example, the following command would restrict the features to CDS features
of at least 500 base pairs.
<blockquote class="call">
p3-all-genomes --eq genome_name,Mycoplasma | p3-get-genome-features --eq feature_type,CDS --ge na_length,500 --attr patric_id --attr product

</blockquote>
</dd>
<dt>p3-get-genome-contigs</dt>
<dd>Given a file of genome IDs, returns the contigs. The following pipeline returns all the contigs in genome 28903.66.
<blockquote class="call">
p3-echo -t genome_id 28903.66 | p3-get-genome-contigs --attr sequence_id --attr sequence

</blockquote>
The output will have three columns, including the genome ID, the ID of the contig, and the actual DNA sequence (which can be quite long).
Again, use the <code>--fields</code> option to see which fields are available for output and filtering in the contig records.
</dd>
<dt>p3-get-genome-drugs</dt>
<dd>Given a file of genome IDs, output the drug resistance information we have on those genomes. For many genomes, no such data is yet
available, so it is not hard to get an empty file output from this command. The following pipeline outputs the default resistance
data information for all the Acinetobacter genomes.
<blockquote class="call">
p3-all-genomes --eq genome_name,Acinetobacter | p3-get-genome-drugs
</blockquote>
<blockquote class="resp">
genome.genome_id    genome_drug.genome_id   genome_drug.antibiotic  genome_drug.resistant_phenotype
1409922.5   1409922.5   amikacin    Resistant
1409922.5   1409922.5   gentamicin  Resistant
1409922.5   1409922.5   tobramycin  Resistant
1409922.5   1409922.5   imipenem    Resistant
1409922.5   1409922.5   ceftazidime Resistant
1409922.5   1409922.5   ceftriaxone Resistant
1409922.5   1409922.5   aztreonam   Resistant
1409922.5   1409922.5   ciprofloxacin   Resistant
1409922.5   1409922.5   levofloxacin    Resistant
1409922.5   1409922.5   tetracycline    Resistant
1409922.5   1409922.5   cefepime    Intermediate
1409923.3   1409923.3   amikacin    Resistant
1409923.3   1409923.3   gentamicin  Resistant

</blockquote>
<dt>p3-all-drugs</dt>
<dd>This script lists anti-microbial drugs from the database. Use the <code>--fields</code> option to see a list of all the
fields you can select. The default is to simply list the antibiotic name, as shown below.
<blockquote class="call">
p3-all-drugs

</blockquote>
<blockquote class="resp">
drug.antibiotic_name
amikacin
amoxicillin
amoxicillin/clavulanic acid
ampicillin
ampicillin/sulbactam
azithromycin
aztreonam
bacitracin
capreomycin
cefaclor
cefazolin

</blockquote>
</dd>
<dt>p3-get-drug-genomes</dt>
<dd>Given a file of antibiotic names, display all the resistance data for those antibiotics. The following pipeline lists all the
genomes resistant to at least one drug.
<blockquote class="call">
p3-all-drugs | p3-get-drug-genomes --attr genome_id --attr genome_name --eq resistance_phenotype,resistant

</blockquote>
<blockquote class="resp">
drug.antibiotic_name    genome_drug.genome_id   genome_drug.genome_name
amikacin    1304920.3   Klebsiella pneumoniae 361_1301
amikacin    1427177.3   Mycobacterium tuberculosis XTB13-081
amikacin    1427178.3   Mycobacterium tuberculosis XTB13-082
amikacin    1427180.3   Mycobacterium tuberculosis XTB13-084
amikacin    1427185.3   Mycobacterium tuberculosis XTB13-091
amikacin    1427191.3   Mycobacterium tuberculosis XTB13-097
amikacin    1427192.3   Mycobacterium tuberculosis XTB13-098
amikacin    1427193.3   Mycobacterium tuberculosis XTB13-100
amikacin    1427199.3   Mycobacterium tuberculosis XTB13-107
amikacin    1427200.3   Mycobacterium tuberculosis XTB13-108
amikacin    1427202.3   Mycobacterium tuberculosis XTB13-110
amikacin    1427204.3   Mycobacterium tuberculosis XTB13-112
amikacin    1427207.3   Mycobacterium tuberculosis XTB13-115

</blockquote>
</dd>
<dt>p3-get-family-features</dt>
<dd>Given a list of protein family IDs, get all the features in the families. PATRIC supports three types of protein families--
<code>local</code>, <code>global</code>, and <code>figfam</code>. The <code>--ftype</code> parameter specifies the type of family
desired. So, for example, the following pipeline finds the global family for the feature <i>fig|1105121.3.peg.460</i> and
then lists the ID and product of each family member.
<blockquote class="call">
p3-echo -t feature_id "fig|1105121.3.peg.460" | p3-get-feature-data --attr pgfam_id | p3-get-family-features --ftype=global --attr patric_id --attr product

</blockquote>
Note that the features found are listed in the column <b>feature.patric_id</b>, while the original feature is maintained in the first
column <b>feature_id</b>.
<blockquote class="resp">
feature_id  feature.pgfam_id    feature.patric_id   feature.product
fig|1105121.3.peg.460   PGF_00112374    fig|1313.8637.peg.2087  hypothetical protein
fig|1105121.3.peg.460   PGF_00112374    fig|1313.8636.peg.1563  hypothetical protein
fig|1105121.3.peg.460   PGF_00112374    fig|1313.8645.peg.110   hypothetical protein
fig|1105121.3.peg.460   PGF_00112374    fig|1313.12423.peg.2037 hypothetical protein
fig|1105121.3.peg.460   PGF_00112374    fig|1330044.3.peg.533   hypothetical protein
fig|1105121.3.peg.460   PGF_00112374    fig|1313.5699.peg.1778  hypothetical protein
fig|1105121.3.peg.460   PGF_00112374    fig|1313.5750.peg.307   hypothetical protein
fig|1105121.3.peg.460   PGF_00112374    fig|1313.5754.peg.739   hypothetical protein
fig|1105121.3.peg.460   PGF_00112374    fig|1313.5758.peg.1823  hypothetical protein
fig|1105121.3.peg.460   PGF_00112374    fig|1313.5781.peg.1819  hypothetical protein
fig|1105121.3.peg.460   PGF_00112374    fig|1313.5778.peg.686   hypothetical protein
fig|1105121.3.peg.460   PGF_00112374    fig|1313.5729.peg.1554  hypothetical protein

</blockquote>
</dd>
<dt>p3-get-feature-data</dt>
<dd>Given a file of feature IDs, return data from those features. Again, use the <code>--fields</code> option to list
the fields you can use for filtering and display. The following pipeline lists the function (product) and protein sequence of each peg of less
than 300 base pairs in the genome 1105121.3.
<blockquote class="call">
p3-echo -t genome_id 1105121.3 | p3-get-genome-features --lt na_length,300 --eq feature_type,CDS --attr patric_id | p3-get-feature-data --attr product --attr aa_sequence

</blockquote>
<blockquote class="resp">
genome_id   feature.patric_id   feature.product feature.aa_sequence
1105121.3   fig|1105121.3.peg.1487  BOX elements    MKIKEQTRKLAASCSKHCFEVVDKTDEVSYIYNPRRR
1105121.3   fig|1105121.3.peg.1508  hypothetical protein    MISTTYRNHRKRFGLRMNLIAEKVSKTLDKTFDKDVREIPTSQFYQKFVDEMGRTYSGNLILQELITVNGAYKATYIGELSSN
1105121.3   fig|1105121.3.peg.1557  hypothetical protein    MKREVISNGNDGPSQEILIFTKQIRHWILSDQVISGKRKLFFREDTPKEILDLYENIKSKLDFAYQEVYSNNGLKKYEK
1105121.3   fig|1105121.3.peg.1598  BOX elements    MKIKEQTRKLAAGCSKHCFEVVDRTDEVSNLHTARRR
1105121.3   fig|1105121.3.peg.1776  hypothetical protein    MVASASASSTSTQAQEQVDKSELRALSQELDQRLKALATVSDPKIDATKAVLLDAQKAPEDSALTE
1105121.3   fig|1105121.3.peg.10    hypothetical protein    MENLLDVIEQFLGLSDEKLEELADKNQLLRLQEEKERKNA
1105121.3   fig|1105121.3.peg.94    BOX elements    MKIKEQTRKLAAGCSKHCFEVVDKTDEVSYIYLRQGEADAV
1105121.3   fig|1105121.3.peg.220   Ribonucleotide reductase of class III (anaerobic), large subunit (EC 1.17.4.2)  MVKRTCGYLGNPQARPMVNGRHKEIAARVKHMNGSTIKIAGHQVTN
1105121.3   fig|1105121.3.peg.228   LSU ribosomal protein L23p (L23Ae)  MNLYDVIKKPVITESSMAQLEAGKYVFEVDTRAHKLLIKQAVEAAFEGVKVANVNTINVKPKAKRVGRYTGFTNKTKKAIITLTADSKAIELFAAEAE
1105121.3   fig|1105121.3.peg.230   SSU ribosomal protein S19p (S15e)   MGRSLKKGPFVDEHLMKKVEAQANDEKKKVIKTWSRRSTIFPSFIGYTIAVYDGRKHVPVYIQEDMVGHKLGEFAPTRTYKGHAADDKKTRRK

</blockquote>
</dl>
<h2>What Is a PATRIC Workspace?</h2>

<p>Users of PATRIC have access to a wealth of public data that support
interpretation of prokaryotic genomes.  The PATRIC team actively integrates newly-sequenced
genomes, data relating to antimicrobial resistance, expression data, pathway data and subsystem data
into an integrated framework that can be queried using either the PATRIC UI or the CLI.</p>

<p>In the PATRIC UI, your workspace looks a lot like a standard file system, divided into folders full of
data. In addition to files you upload, such as FASTA and FASTQ files, there will also be typed objects
such as genomes, feature groups, and genome groups. The CLI allows you to move these typed objects between
your workspace and your file system so you can manipulate them at will.</p>

<h3>Logging In</h3>

<p>To access your workspace, you need a PATRIC account. If you do not have one already, go to
<a href="https://user.patricbrc.org/register">https://user.patricbrc.org/register</a> and register
now.</p>

<p>Now that you have a working user name and password, you can use the <b>p3-login</b> script to
tell the CLI who you are. For example, if your name is <code>rastuser25</code>, you would type</p>

<blockquote class="call">
p3-login rastuser25

</blockquote>

<p>The script asks you for your password and places a special file on your hard drive that can be used
to get authorized access to your workspace data. To log out again, simply use</p>

<blockquote class="call">
p3-login --logout

</blockquote>


<p>At any time, you can verify your login status using</p>

<blockquote class="call">
p3-login --status

</blockquote>

<p>If you are logged out, it will respond</p>

<blockquote class="resp">
You are currently logged out of PATRIC.

</blockquote>

<p>If you are logged in, you will get something like</p>

<blockquote class="resp">
You are logged in as rastuser25@patricbrc.org.

</blockquote>

<h3>Working with Genome Groups</h3>

<p>Your workspace looks like a full-blown file system, but there are three special folders.</p>
<ul>
<li><b>Genome Groups</b> contains named lists of genomes.</li>
<li><b>Feature Groups</b> contains named lists of features.</li>
<li><b>QuickData</b> contains folders full of genomes you submitted through the CLI annotation interface.</li>
</ul>

<p>To create a genome group, you use <b>p3-put-genome-group</b>. Say, for example, you want to examine
Streptococcus penumoniae genomes that are resistant to penicillin. The following query command will return this list
of genomes (we will discuss all query commands in more details later).</p>

<blockquote class="call">
p3-echo -t antibiotic penicillin | p3-get-drug-genomes --eq "genome_name,Streptococcus pneumoniae"  --eq resistant_phenotype,resistant --attr genome_id --attr genome_name &gt;resist.tbl

</blockquote>

<p>This particular
command asks for data from the anti-microbial resistance table. Each record in this table posits a relationship
between a genome and an antibiotic drug. We are accessing the table from the direction of taking a drug and finding
resistant genomes. To do this, we need a file with a drug name in it. The <b>p3-echo</b> command creates this file:
the <code>-t antibiotic</code> parameter tells it we want a one-column file with a column header of <code>antibiotic</code>.
We put the single record <code>penicillin</code> in that column.<p>
<p>The antibiotic file is then piped into <b>p3-get-drug-genomes</b>. Its parameters do the following.</p>
<dl>
<dt><code>--eq "genome_name,Streptococcus pneumoniae"</code></dt>
<dd>Only include records for Streptococcus pneumoniae genomes. Because this is a string field, it does a substring match.
A genome name including follow-on strain information (e.g. <code>Streptococcus pneumoniae strain LMG2888</code>)
will still match.</dd>
<dt><code>--eq resistant_phenotype,resistant</code></dt>
<dd>Only include records that state the genome is resistant.</dd>
<dt><code>--attr genome_id</code></dt>
<dd>Output the genome ID.</dd>
<dt><code>--attr genome_name</code></dt>
<dd>Output the genome name.</dd>
</dl>

<p>When the command completes, the file <b>resist.tbl</b> will contain around 114 lines beginning with
the following.</p>

<blockquote class="resp">
antibiotic   genome_drug.genome_id   genome_drug.genome_name
penicillin   1313.7006   Streptococcus pneumoniae P310010-154
penicillin   1313.7016   Streptococcus pneumoniae P310937-212
penicillin   1313.7018   Streptococcus pneumoniae P311313-217
penicillin   760749.3    Streptococcus pneumoniae GA05248
penicillin   760763.3    Streptococcus pneumoniae GA11304
penicillin   760765.3    Streptococcus pneumoniae GA11663
penicillin   760766.3    Streptococcus pneumoniae GA11856
penicillin   760769.3    Streptococcus pneumoniae GA13338
penicillin   760771.3    Streptococcus pneumoniae GA13455
penicillin   760776.3    Streptococcus pneumoniae GA14373
penicillin   760777.3    Streptococcus pneumoniae GA14688

</blockquote>

<p>Now we want to create a group for these genomes called <b>resist_strep</b>. We use the following command.</p>

<blockquote class="call">
p3-put-genome-group --col=2 resist_strep &lt;resist.tbl

</blockquote>

<p>The <code>--col=2</code> tells the command that the genome IDs are in the second column. The genome group is simply
a set of genome IDs, so the other columns will be ignored by the command. You can read the group back at any time
using <b>p3-get-genome-group</b>.</p>

<blockquote class="call">
p3-get-genome-group resist_strep

</blockquote>

<p>Will output</p>

<blockquote class="resp">
resist_strep.genome_id
1313.7006
1313.7016
1313.7018
760749.3
760763.3
760765.3
760766.3
760769.3
760771.3

</blockquote>

<p>and so on. Note that if you want to see the names as well, you can use the <b>p3-get-genome-data</b> command to add them</p>

<blockquote class="call">
p3-get-genome-group resist_strep | p3-get-genome-data --attr genome_name

</blockquote>

<blockquote class="resp">
resist_strep.genome_id  genome.genome_name
1313.7006   Streptococcus pneumoniae P310010-154
1313.7016   Streptococcus pneumoniae P310937-212
1313.7018   Streptococcus pneumoniae P311313-217
760749.3    Streptococcus pneumoniae GA05248
760763.3    Streptococcus pneumoniae GA11304
760765.3    Streptococcus pneumoniae GA11663
760766.3    Streptococcus pneumoniae GA11856
760769.3    Streptococcus pneumoniae GA13338
760771.3    Streptococcus pneumoniae GA13455
760776.3    Streptococcus pneumoniae GA14373
760777.3    Streptococcus pneumoniae GA14688

</blockquote>

<p>Next we will ask for the genomes that are susceptible to penicillin. We use the same command as before except we put
<code>susceptible</code> in place of <code>resistant</code>. We're going to pipe the results directly into
<b>p3-put-genome-group</b> to store them in our workspace.</p>

<blockquote class="call">
p3-echo -t antibiotic penicillin | p3-get-drug-genomes --eq "genome_name,Streptococcus pneumoniae"  --eq resistant_phenotype,susceptible --attr genome_id --attr genome_name | p3-put-genome-group --col=2 weak_strep

</blockquote>

<p>Now when you ask for the group back, you would get something like the following.

<blockquote class="call">
p3-get-genome-group weak_strep

</blockquote>

<blockquote class="resp">
weak_strep.genome_id
1313.6939
1313.6941
1313.6942
1313.6944
1313.6947
1313.7001
1313.7002
1313.7007
1313.7009

</blockquote>

<h3>Working with Feature Groups</h3>

<p>We want to look at features in the resistant Streptococcus pneumoniae genomes that distinguish them from the susceptible ones.
Then we will gather those features into a feature group and store them in our workspace so we can work with them later.
We already have our resistant group in a file called <i>resist.tbl</i>. We will use the <b>p3-get-genome-group</b>
command to pull the susceptible genomes out of our workspace and pipe it into <b>p3-signature-families</b>, which
counts and compares protein families.

<blockquote class="call">
p3-get-genome-group weak_strep | p3-signature-families --gs1=resist.tbl >families.tbl

</blockquote>

<p>The <b>p3-signature-families</b> script asks for two sets of genome IDs as input,
specified as file names using the command-line parameters <b>--gs1</b> (interesting set) and <b>--gs2</b> (other set). The output
contains protein families that are common in the interesting set but not in the other set. (For convenience, any genomes in the
interesting set are removed from the other set before processing.) If a set file is not specified, it is taken from the standard
input. In this case, that would be the other set, since there is no <b>--gs2</b> parameter. That is piped in from
the <b>p3-get-genome-group</b> command.</p>

<p>Our signature families analysis script yields four protein families, displayed in the following output.</p>

<blockquote class="resp">
counts_in_set1  counts_in_set2  family.family_id    family.product
92  10  PGF_00112374    hypothetical protein
92  10  PGF_00303700    hypothetical protein
92  10  PGF_03497231    hypothetical protein
91  10  PGF_03497236    hypothetical protein

</blockquote>

<p>The next step is to convert the families into feature IDs. The <b>p3-get-family-features</b> script performs that
function. We will use the following command.</p>

<blockquote class="call">
p3-get-family-features --gFile=resist.tbl --gCol=1 --ftype=global --col=family.family_id &lt;families.tbl

</blockquote>

<p>The parameters work as follows.
<dl>
<dt>--gFile=resist.tbl</dt>
<dd>Only features from the genomes listed in the file <i>resist.tbl</i> should be included in the output.</dd>
<dt>--gCol=1</dt>
<dd>The genome IDs in <i>resist.tbl</i> are in the first column.</dd>
<dt>--ftype=global</dt>
<dd>The family IDs in the input file are global families. (Local families and FIGfams are also supported, but
the <b>p3-signature-families</b> script uses global families.)</dd>
<dt>--col=family.family_id</dt>
<dd>The protein family IDs in the input file are in a column named <code>family.family_id</code>.</dd>
</dl>

<p>The output looks something like this.</p>

<blockquote class="resp">
counts_in_set1  counts_in_set2  family.family_id    family.product  feature.patric_id   feature.feature_type    feature.location    feature.product
92  10  PGF_00112374    hypothetical protein    fig|1313.5465.peg.1094  CDS complement(6450..9101)  hypothetical protein
92  10  PGF_00112374    hypothetical protein    fig|1313.5418.peg.2058  CDS complement(6450..9101)  hypothetical protein
92  10  PGF_00112374    hypothetical protein    fig|1313.5645.peg.1124  CDS 3518..6169  hypothetical protein
92  10  PGF_00112374    hypothetical protein    fig|1069623.3.peg.450   CDS 391076..393727  hypothetical protein
92  10  PGF_00112374    hypothetical protein    fig|1313.6729.peg.1445  CDS 32402..35053    hypothetical protein
92  10  PGF_00112374    hypothetical protein    fig|1313.6728.peg.1682  CDS complement(6420..9071)  hypothetical protein
92  10  PGF_00112374    hypothetical protein    fig|1313.6723.peg.2045  CDS 297..2948   hypothetical protein
92  10  PGF_00112374    hypothetical protein    fig|760766.3.peg.438    CDS 293743..296403  hypothetical protein
92  10  PGF_00112374    hypothetical protein    fig|1313.6724.peg.811   CDS 32400..35051    hypothetical protein

</blockquote>

<p>We didn't tell <b>p3-get-family-features</b> what attributes of the features to display, so it defaulted to ID, type,
location, and functional assignment (product). We don't have time to examine these features in detail now, but we can put them
in a feature group by piping them into <b>p3-put-feature-group</b> as follows.</p>

<blockquote class="call">
p3-get-family-features --gFile=resist.tbl --gCol=1 --ftype=global --col=family.family_id &lt;families.tbl | p3-put-feature-group --col=feature.patric_id resist_fids
</blockquote>

<p>In the <b>p3-put-feature-group</b> command, the <code>--col=feature.patric_id</code> parameter tells the command
that the feature IDs are in the column with that heading, and <code>resist_fids</code> is the group name. When you
decide to examine the features in greater detail, you can pull back the feature IDs using <b>p3-get-feature-group</b>.</p>

<blockquote class="call">
p3-get-feature-group resist_fids

</blockquote>

<p>The output will look something like this.</p>

<blockquote class="resp">
resist_fids.patric_id
fig|1105121.3.peg.460
fig|1069626.3.peg.432
fig|1313.6771.peg.1961
fig|1313.5503.peg.1279
fig|1313.5634.peg.1224
fig|1069624.3.peg.437
fig|1069628.3.peg.451
fig|1313.5669.peg.1163
fig|1313.6725.peg.2115

</blockquote>

<p>At any time, you can get a complete list of the groups in your workspace using the <b>p3-list-genome-groups</b>
command or the <b>p3-list-feature-groups</b> command. So, if you have been following along the above examples and
your workspace was empty before you began, you would see the following.</p>

<blockquote class="call">
p3-list-genome-groups

</blockquote>

<blockquote class="resp">
resist_strep
weak_strep

</blockquote>

<blockquote class="call">
p3-list-feature-groups

</blockquote>

<blockquote class="resp">
resist_fids

</blockquote>

<h2>Using PATRIC Apps: Assembling Reads and Annotating Genomes</h2>

<p>There are a growing number of PATRIC apps, and you will be able to use many of them via the CLI.
To illustrate how, we show you how to assemble collections of reads into contig, and how to annotate one
or more genomes (hundreds at a time, if you wish).</p>

<h3>Moving Data To and From Your PATRIC Workspace</h3>

<h3> Assembling Reads Into Contigs </h3>

<p>The assembly app allows you to take one or more files of DNA <i>reads</i>
and produce an output file containing DNA <i>contigs</i>.  The exact steps one would use
to assemble reads are as follows:</p>
<ol>
<li> Upload the one or more files containing reads to your workspace:
<pre>
  example commands following exactly what is shown in the previous section
</pre></li>
<li> Now use....
<pre>
  more example commands
</pre></li>
</ol>

<h3> Annotating Genomes</h3>

<p>The PATRIC app that can be used to annotate a genome takes as input</p>
<ol>
<li>a file of contigs,</li>
<li>the genetic code used by the organism (use "4" if you are
annotating a <i>Mycoplasma</i> and "11", otherwise).  See
<a>link to NCBI</a> for details.
<li>The taxonomy ID you wish to assign to the genome (see <a>taxonomy thing</a> for details)</li>
</ol>

<p>It produces as output a <i>Genome Type Object (GTO)</i>.  This object
contains all of the data created during the annotation.  You can download a GTO
to your machine and then use p3-scripts to extract data from the GTO.
Here is an example of how to download the output GTO and get a listing of
the features identified in the annotations.</p>

<pre>
    downloading example
</pre>

<h1>Extracting Data from PATRIC</h1>

<p>You can do a great deal more with the CLI than just run PATRIC apps -- you can extract a large and
growing body of data that you may well find useful in studying organisms and their genomes. Above
we listed the five entities that contain PATRIC data and the scripts for navigating them.


<h2>The Entities</h2>

You can think of PATRIC as a collection of data items that describe entities of just a few types.
<i>Genomes</i> and <i>Features</i> are the most commonly accessed types of entities.

<pre>

============== Evolving Notes

</pre>

The P3-scripts interface reflects an underlying ERDB.
The initial implemenation will offer access to the following
entities:
<ul>
<li>Genome
<li>Feature
<li>ProteinFamily
<li>Role
<li>Subsystem
<li>PopSys (a row in a populated subsystem)
<li>Drug
</ul>

<h3>Accessing Data from an Entity (set)</h3>

For each entity we support something like
<pre>
  p3-get-Genome-data SearchFor FieldsToExtract

      id is always default field

</pre>
Here <i>SearchFor</i> encodes a set of search contraints.
<i>FieldsToExtract</i> encodes a list of desired attributes.
For important relationships, we use things like
<pre>
  p3-Genome2Drug SearchFor FieldsToExtract
</pre>

<h2>Tables, Commands, and Headers</h2>

Many of the P3-scripts are commands that take a tab-separated table with column headers
as input and produce an altered table as output.  The decision to use column headers is a
deviation from the tools we built as SVR and SVC scripts.  It has both benefits and costs.
For the unix commands our use of headers introduces surmountable complexities.  For example,
we will implement <b>P3-sort</b> as a command that reads and prints the header line, and then uses
<br>
<pre>
  tail ``-n+2'' | sort
</pre>
to print the sorted contexts of the file.  There will be P3-scripts corresponding to at
least the following UNIX tools:
<ul>
<li>echo (p3-echo)
<li>head (p3-head)
<li>tail (p3-tail)
<li>grep (p3-grep)
<li>sort (p3-sort)
<li>cut  (p3-cut)
</ul>
The Unix <i>cut</i> does not let you establish the order of the retained columns.
p3-cut (which is a renamed version of svc_extract) will.


<h2>Classes of Commands that Need to be Cast as P3-scripts</h2>

<h3>Access to the PATRIC Workspaces</h3>

We have a set of commands that support creation and maintenance of PATRIC workspaces.

<h3>Running PATRIC apps and Retrieving the Ouput</h3>

We support the ability to run apps and retrieve the output.

<h3>Extracting, Copying, and Mining Genome Typed Objects (GTOs) </h3>

<h2>Queries that Need to be Supported</h2>
Note that some of the queries could probably best be handled using
code from someone else.  That would, of course, be fine with me.

<ol>
<li>Which (how many) Streptococcus genomes do we have?
<pre>
p3-all-genomes --equal genus,Streptococcus --count

genome.count
11475
</pre>

<li>Which Streptococcus genomes do we have?

<pre>
  p3-all-genomes --equal genus,Streptococcus --attr genome_name --attr genome_length
</pre>
<li>What fraction of the genomes in genus G are resistant to drug D?
<pre>
  p3-all-genomes --equal genome_name,Staphylococcus â€“count

genome.count
10634

p3-echo -t antibiotic methicillin | p3-get-drug-genomes --equal resistant_phenotype,resistant --equal genome_name,Staphylococcus --count

antibiotic   genome_drug.count
methicillin  1064
</pre>

<li>What fraction of the genomes in genus X have an occurrence of role R?
<li>What fraction of the genomes in genus X have multiple occurrences of role R?
<li>What fraction of the genomes in the populated subsystem S are Mycobacterium genomes?
<li>What fraction of the Mycobacterium genomes in the populated subsystem S have pegs that connect to role R?
<li>Which subsystems are active in genome G?
<li>Which pegs implement role X in genome G?
<li>Which subsystems are active in fewer than 5 genomes?
<li>How many genomes have active occurrences of subsystem S?
<li>Which roles of subsystem S are connected to pegs?
<li>Which protein family is peg P in?
<pre>
Which protein family is fig|46170.310.peg.738 in?

p3-echo -t feature.feature_id "fig|46170.310.peg.738" | p3-get-feature-data --attr pgfam_id

feature.feature_id feature.pgfam_id
fig|46170.310.peg.738 PGF_00040464

</pre>

<li>Which roles does peg P implement?
<pre>
Which roles does fig|46170.310.prh.738 implement?

p3-echo -t feature.feature_id "fig|46170.310.peg.738" | p3-get-feature-data --attr product

feature.feature_id      feature.product
fig|46170.310.peg.738   Putative cysteine desulfurase, associated with tRNA 4-thiouridine synthase
</pre>

<li>What drugs is genome G resistant to?

What drugs is 46170.310 resistant to?
<pre>
p3-echo -t genome.genome_id 46170.310 | p3-get-genome-drugs --equal resistant_phenotype,resistant  --attr antibiotic

genome.genome_id        genome_drug.antibiotic
46170.310       ciprofloxacin
46170.310       erythromycin
46170.310       gentamicin
46170.310       methicillin
46170.310       penicillin
46170.310       trimethoprim/sulfamethoxazole
</pre>

<li>What genomes are resistant to drug D?

What genomes are resistant to erythromycin?
<pre>
p3-echo -t antibiotic,erythromycin | p3-get-drug-genomes --equal resistant_phenotype,resistant --attr genome_id --attr genome_name

antibiotic    genome_drug.genome_id   genome_drug.genome_name
erythromycin  1280.4920     Staphylococcus aureus P210110-35
erythromycin  1280.4930     Staphylococcus aureus P210184-226
erythromycin  1280.4940     Staphylococcus aureus P210369-10
erythromycin  1280.4960     Staphylococcus aureus P210464-28
erythromycin  1280.4970     Staphylococcus aureus P310372-198
erythromycin  1280.4990     Staphylococcus aureus P311202-207
erythromycin  1313.6942     Streptococcus pneumoniae P110340-157
erythromycin  1313.7001     Streptococcus pneumoniae P210774-233

</pre>

<li>Given two pegs from the same contig on a genome, how close are they on the chromsome?
<pre>
     p3-echo -t f1.patric_id -t f2.patric_id "fig|1302.21.peg.966" "fig|1302.21.peg.1019" | p3-feature-gap

     f1.patric_id    f2.patric_id    gap
     fig|1302.21.peg.966     fig|1302.21.peg.1019    55253
</pre>


<li>Are pegs P1 and P2 located within 5 kb of one another on the chromosome?
<pre>
     p3-echo -t f1.patric_id -t f2.patric_id "fig|1302.21.peg.966" "fig|1302.21.peg.1019" | p3-feature-gap

     f1.patric_id    f2.patric_id    gap
     fig|1302.21.peg.966     fig|1302.21.peg.1019    55253
</pre>
<li>Give me a fasta file of the contigs of genome G.
<pre>
    p3-genome-fasta 1302.21
</pre>
<li>Give me the protein sequences for the pegs in genome G.
<pre>
    p3-genome-fasta --protein 1302.21
</pre>
<li>Given a list of genomes L, produce a list of pairs of roles that are implemented
    by pegs (from genomes in L) that are close on the chromosome (along with a count of the number of occurrences of the pair).
<pre>
    p3-get-genome-features --attr sequence_id --attr location --attr product &lt;genomes.tbl | p3-function-to-role | p3-generate-close-roles

</pre>
<li> Given two sets of genomes Gs1 and Gs2, list the roles implemented in genomes in Gs1, but not by
genomes implemeted in Gs2.
<li>What is the average length of proteins in family F, along with the standard deviation.
<li>What genomes in list Gs have GC content values greater than 60%.
<li>For functional roles R1 and R2, give a correlation coefficient for occurrences of pegs implementing
these roles in the entire set of genomes.
<li>For subsystems S1 and S2, compute the correlation coeffient of occurrences of S1 and S2
in a list of genomes.
<li>Compute the minimum number of unique roles in the subsystem S that are implemented by pegs.  That is
    calculate <i>min_pegs</i>, the minimum number of pegs required for an active instance of the subsystems
(at least so far).  Which genomes have only <i>min_pegs</i> in an active instance?
<li>Compute the upstream and downstream regions for pegs in list pegL.
<li>Return features from a designated list of types for a genome G -- sorted by location on the chromosome.
<li>Given a genome G that does not occur in subsystem S, compute the minimum sets of roles needed to
    include all of the roles that occur in at least one active role (there may be multiple ways
    to complete an incomplete set).
<li>Which kmers occur in one list of genomes, but not in a second set?  Here we are defining
signature kmers.
<li>Which roles occur in one list of genomes, but not in a second set?  Here we are defining signature roles.
<li>Which subsystems occur in one list of genomes, but not in a second set?  Here we are defining signature subsystems.
<li>Compute the codon usage for pegs in a specified genome.  This requires getting the DNA sequence
    for the genome, getting the peg translations, and tabulating the codon counts.
<li>Given a possible frameshift (in the form of two adjacent pegs), try to "fix it".  We have a version
of the code to do this, and it involves a Smith-Waterman algorithm.  Hopefully, we can just lift the
existing version.
<li>Given a genome G, get roles that may be split over multiple contigs.
<li>Given a genome G and a close genome G', use G' to look for pegs in G1 that might span contigs.

</ol>

</div>
</body>
</html>