<!doctype html>
<html>
<head>
<meta charset="utf-8" />
<title>Common Tasks With P3 Scripts</title>
<link href="../css/Basic.css" rel="stylesheet" type="text/css">
</head>
<body>
<div id="Dump">
<h1>Common Tasks With P3 Scripts</h1>
<p>In this tutorial, we will present examples of common tasks and show how to accomplish them with P3 scripts.</p>
<dl>
<dt>How many Streptococcus genomes do we have?</dt>
<dd>
<pre class="call">
p3-all-genomes --equal genus,Streptococcus --count
</pre>

<pre class="resp">
genome.count
11836
</pre>
Note the use of the <code>--count</code> command-line option to produce a count of the results instead of the results themselves.
</dd>

<dt>Which Streptococcus genomes do we have?</dt>
<dd>
<pre class="call">
p3-all-genomes --equal genus,Streptococcus --attr genome_name
</pre>
<pre class="resp">
genome.genome_id    genome.genome_name
1302.21 Streptococcus gordonii strain DD07
1303.76 Streptococcus oralis strain DD05
1303.77 Streptococcus oralis strain DD14
1303.78 Streptococcus oralis strain DD15
1303.79 Streptococcus oralis strain DD16
1303.80 Streptococcus oralis strain DD20
1303.81 Streptococcus oralis strain DD21
1303.82 Streptococcus oralis strain DD27
1303.83 Streptococcus oralis strain DD30
</pre>
<b>p3-all-genomes</b> always includes the ID, so all we need for the <code>--attr</code> parameter is the name field. If you
intend to pipe the results into another script, specify the attributes in the order you want them to appear.
<pre class="call">
p3-all-genomes --equal genus,Streptococcus --attr genome_name --attr genome_id
</pre>
<pre class="resp">
genome.genome_name  genome.genome_id
Streptococcus gordonii strain DD07  1302.21
Streptococcus oralis strain DD05    1303.76
Streptococcus oralis strain DD14    1303.77
Streptococcus oralis strain DD15    1303.78
Streptococcus oralis strain DD16    1303.79
Streptococcus oralis strain DD20    1303.80
Streptococcus oralis strain DD21    1303.81
Streptococcus oralis strain DD27    1303.82
Streptococcus oralis strain DD30    1303.83
</pre>
</dd>
<dt>What are all the features in the genome 83333.219?</dt>
<dd>The script <b>p3-get-genome-features</b> retrieves all the features for a genome. In addition to the feature ID, we will want the functional
assignment, which is in the <code>product</code> field. The command pipe is as follows.
<pre class="call">
p3-echo 83333.219 | p3-get-genome-features --attr patric_id --attr product
</pre>
<pre class="resp">
id      feature.patric_id       feature.product
83333.219       fig|83333.219.peg.1     tRNA:Cm32/Um32 methyltransferase
83333.219       fig|83333.219.peg.2     Aspartokinase (EC 2.7.2.4) / Homoserine dehydrogenase (EC 1.1.1.3)
83333.219       fig|83333.219.peg.3     Homoserine kinase (EC 2.7.1.39)
83333.219       fig|83333.219.peg.4     Threonine synthase (EC 4.2.3.1)
83333.219       fig|83333.219.peg.5     Uncharacterized protein YaaX
83333.219       fig|83333.219.peg.6     UPF0246 protein YaaA
83333.219       fig|83333.219.peg.7     Putative alanine/glycine transport protein
83333.219       fig|83333.219.peg.8     Transaldolase (EC 2.2.1.2)
83333.219       fig|83333.219.peg.9     Molybdopterin adenylyltransferase (EC 2.7.7.75)
83333.219       fig|83333.219.peg.10    Succinate-acetate/proton symporter SatP
</pre>
</dd>
<dt>What fraction of the genomes in genus Staphylococcus are resistant to methicillin?</dt>
<dd>This is a two-step process, since we need two numbers.
<pre class="call">
p3-all-genomes --equal genome_name,Staphylococcus --count
</pre>
We isolate the genus by doing a string match on the genome name, since equality for string fields matches if the value is a substring. We could
also use <code>--equal genus,Staphylococcus</code> and get an equivalent result.
<pre class="resp">
genome.count
10716
</pre>
To get the count of resistant genomes, we need to pipe the drug name into <b>p3-get-drug-genomes</b>. Here we don't have the option of
using the <i>genus</i> field, since only the genome name is present in the drug-genome records, not the entire taxonomy.
<pre class="call">
p3-echo -t antibiotic methicillin | p3-get-drug-genomes --resistant --equal genome_name,Staphylococcus --count
</pre>
<pre class="resp">
antibiotic   genome_drug.count
methicillin  1064
</pre>
The answer is 1064 * 100 / 10716 or 9.93%.
</dd>
<dt>
<dt>Which global protein family is fig|46170.310.peg.738 in?</dt>
<dd>
<pre class="call">
p3-echo -t patric_id "fig|46170.310.peg.738" | p3-get-feature-data --attr pgfam_id
</pre>
There are a couple of important things here. We use the <i>pgfam_id</i> field to get the global protein family (<i>plfam_id</i> would be
used to get the local protein family). Also, the feature ID is enclosed in double quotes on the command line so that the vertical bar doesn't confuse
the command-line shell.
<pre class="resp">
patric_id feature.pgfam_id
fig|46170.310.peg.738 PGF_00040464
</pre>
</dd>
<dt>What function does fig|46170.310.peg.738 implement?</dt>
<dd>
The function is stored in the feature table's <i>product</i> attribute.
<pre class="call">
p3-echo -t patric_id "fig|46170.310.peg.738" | p3-get-feature-data --attr product
</pre>
<pre class="resp">
patric_id      feature.product
fig|46170.310.peg.738   Putative cysteine desulfurase, associated with tRNA 4-thiouridine synthase
</pre>
Of course, you could ask this question of several features with a single pipe.
<pre class="call">
p3-echo -t patric_id "fig|46170.310.peg.738" "fig|1313.7001.peg.1189" "fig|66976.18.peg.131" | p3-get-feature-data --attr product
</pre>
<pre class="resp">
patric_id   feature.product
fig|46170.310.peg.738   Putative cysteine desulfurase, associated with tRNA 4-thiouridine synthase
fig|1313.7001.peg.1189  IMP cyclohydrolase (EC 3.5.4.10) / Phosphoribosylaminoimidazolecarboxamide formyltransferase (EC 2.1.2.3)
fig|66976.18.peg.131    hypothetical protein
</pre>
The <b>p3-echo</b> command uses the <code>--title</code> command-line option to determine how many parameters to put on each output
line. Since our example has only one title, the output file has only a single column, and it can be easily piped to <b>p3-get-feature-data</b>.
</dd>
<dt>What drugs is 46170.310 resistant to?</dt>
<dd>
The drug name is in the <i>antibiotic</i> attribute of the genome-drug table. We start with a genome ID and use <b>p3-get-genome-drugs</b>.
<pre class="call">
p3-echo -t genome_id 46170.310 | p3-get-genome-drugs --resistant --attr antibiotic
</pre>
<pre class="resp">
genome_id   genome_drug.antibiotic
46170.310   ciprofloxacin
46170.310   erythromycin
46170.310   gentamicin
46170.310   methicillin
46170.310   penicillin
46170.310   trimethoprim/sulfamethoxazole
</pre>
</dd>
<dt>What genomes are resistant to erythromycin?</dt>
<dd>Here we start with a drug name and use <b>p3-get-drug-genomes</b> to get the genome data.
<pre class="call">
p3-echo -t antibiotic erythromycin | p3-get-drug-genomes --resistant --attr genome_id --attr genome_name
</pre>
<pre class="resp">
antibiotic  genome_drug.genome_id   genome_drug.genome_name
erythromycin    1280.4920   Staphylococcus aureus P210110-35
erythromycin    1280.4930   Staphylococcus aureus P210184-226
erythromycin    1280.4940   Staphylococcus aureus P210369-10
erythromycin    1280.4960   Staphylococcus aureus P210464-28
erythromycin    1280.4970   Staphylococcus aureus P310372-198
erythromycin    1280.4990   Staphylococcus aureus P311202-207
erythromycin    1313.6942   Streptococcus pneumoniae P110340-157
erythromycin    1313.7001   Streptococcus pneumoniae P210774-233
erythromycin    1313.7002   Streptococcus pneumoniae P210824-213
erythromycin    1313.7006   Streptococcus pneumoniae P310010-154
erythromycin    1313.7013   Streptococcus pneumoniae P310795-191
</pre>
</dd>
<dt>How close are fig|1302.21.peg.966 and fig|1302.21.peg.1019 on the chromosome?</dt>
<dd>The script <b>p3-feature-gap</b> gives us this information. Since it expects two feature IDs
on the same input line, we use a <b>p3-echo</b> with two titles to put its two parameters on a single line.
<pre class="call">
p3-echo -t f1.patric_id -t f2.patric_id "fig|1302.21.peg.966" "fig|1302.21.peg.1019" | p3-feature-gap
</pre>
<pre class="resp">
f1.patric_id    f2.patric_id    gap
fig|1302.21.peg.966 fig|1302.21.peg.1019    55253
</pre>
Note that if the features are on different contigs, we get a very high number.
<pre class="call">
p3-echo -t f1.patric_id -t f2.patric_id "fig|1313.7001.peg.1159" "fig|1313.7001.peg.1384" | p3-feature-gap
</pre>
<pre class="resp">
f1.patric_id    f2.patric_id    gap
fig|1313.7001.peg.1159  fig|1313.7001.peg.1384  2000000000
</pre>
The very high number makes it easier to simply compare the distance outputs from <b>p3-feature-gap</b>. Features on
different contigs will always sort as further apart than features on the same contig.
</dd>
<dt>Give me a fasta file of the contigs of genome 1302.21</dt>
<dd>
<pre class="call">
p3-genome-fasta 1302.21
</pre>
<pre class="resp">
>1302.21.con.0001 contig
agctcagttggtagtagcgcatgactgttaatcatgatgtcgtaggttcgagtcctactg
ccggagttatatctataagtaagacaagaaattcttgtctttttatatttattgtgtttt
tgcaatttaatttttaagttcttatttaataaaaagcttgaagattattcttcaagcttt
ttatgtttattaaagaatgcttcatagagggctttaatagctgctttttcttgttcagag
tttactacgagcatgatagaaacttcgctagatccttgagagatcatttgaatattaatt
ttgctgtctgatagagcctttgtagccgtagcagtcagaccgatatgacttttcatttgc
</pre>
</dd>
<dt>Give me the protein sequences for the pegs in genome 1302.21.</dt>
<dd>
<pre class="call">
p3-genome-fasta --protein 1302.21
</pre>
<pre class="resp">
>fig|1302.21.peg.966 putative Zn-dependent protease
MRFLLNLFRFIWRMFWRLVWAGIVAFIILVSVLYLTNPSQTGLTAVRQAVQTAVNQLDTF
LDQQGIHTGLGQNVQNLGEHLTDQHVASSDGARWENARATVYIETENSTFRAAYQEAIKS
WNATGAFTFQLVEDKSQANIIATEMNDSTITAAGEAESQTNVLTKRFTKVTVRLNAYYLL
NNYYGYSHERIVNTASHELGHAIGLDHNESESVMQSAGSFYSIQPIDIQAVKELYQD
>fig|1302.21.peg.969 Putative metallopeptidase (Zinc) SprT family
MNLNEYIKQVSLEDFGWEFRHQAFWNKRLRTTGGRFFPKDGHLDFNPKIYETFGLETFRK
IVRHELAHYHLYYQGKGYRHKDRDFKELLKQVGGLRYAPGLPAKKLKLHYQCRSCCTDFY
RQRRIEIKKYRCGRCKGKLRLLKQER
</pre>
</dd>
<dt>Given a list of genomes, produce a list of pairs of roles that are implemented
by pegs that are close on the chromosome, sorted by number of occurrences.</dt>
<dd>Here we assume our list of genomes is in the file <b>genomes.tbl</b>. The content of this file is shown below.
<pre class="call">
genome_id
1310696.14
66976.17
91890.5
316273.25
186497.12
1353158.3
135461.13
1173954.3
1176728.3
</pre>
We use <b>p3-get-genome-features</b> to get the feature and location data, <b>p3-function-to-role</b> to convert the functions
to roles, and <b>p3-generate-close-roles</b> to compute the physically close roles. Because we only want protein-encoding genes
(pegs), we filter the genome features by type. (If we didn't do this, the output would start with a whole bunch of
generic roles involving ribosomes and CRISPR repeats.) The output is automatically sorted by decreasing number of occurrences.
<pre class="call">
p3-get-genome-features --eq feature_type,CDS --attr sequence_id --attr location --attr product &lt;genomes.tbl | p3-function-to-role | p3-generate-close-roles
</pre>
<pre class="resp">
role1   role2   count
Transposase, IS3/IS911 family   Mobile element protein  33
Mobile element protein  Mobile element protein  29
Lead, cadmium, zinc and mercury transporting ATPase (EC 3.6.3.3) (EC 3.6.3.5)   Copper-translocating P-type ATPase (EC 3.6.3.4) 25
Potassium efflux system KefA protein    Small-conductance mechanosensitive channel  13
Cobalt-zinc-cadmium resistance protein CzcA Cation efflux system protein CusA   13
Gamma-glutamyltranspeptidase (EC 2.3.2.2)   Glutathione hydrolase (EC 3.4.19.13)    13
Efflux ABC transporter, ATP-binding protein Efflux ABC transporter, permease protein    11
</pre>
Note that the occurrence counts are shown in the last column of the output.
</dd>
<dt>What genomes in a list have GC content values greater than 60%.</dt>
<dd>
For this exercise we will use the <b>genomes.tbl</b> file as input.
<pre class="call">
genome_id
1310696.14
66976.17
91890.5
316273.25
186497.12
1353158.3
135461.13
1173954.3
1176728.3
</pre>
The GC content percentage is found in the <i>gc_content</i> attribute, as shown in the example below (we use <b>p3-sort</b>
to sort the results by the content percentage).
<pre class="call">
p3-get-genome-data --attr gc_content --attr genome_name &lt;genomes.tbl | p3-sort gc_content/n
</pre>
<pre class="resp">
genome_id       genome.gc_content       genome.genome_name
91890.5 38.19   Legionella pneumophila subsp. pascullei strain D-7158
66976.17        38.28   Legionella pneumophila serogroup 1 strain Lp01_666
186497.12       40.8    Pyrococcus furiosus DSM 3638
1353158.3       43.41   Methanococcoides vulcani strain SLH 33
135461.13       43.88   Bacillus subtilis subsp. subtilis strain BSD-2
1173954.3       45.1    Vibrio parahaemolyticus O4:K12 str. K1203
1176728.3       50.67   Escherichia coli K71
316273.25       64.56   Xanthomonas campestris pv. vesicatoria str. 85-10
</pre>
As you can see, there is only one genome in this set with a GC content over 60%. To get only that genome, we use the
<code>--gt</code> parameter to filter for specific values of that field.
<pre class="call">
p3-get-genome-data --attr gc_content --attr genome_name --gt gc_content,60 &lt;genomes.tbl
</pre>
<pre class="resp">
genome_id   genome.gc_content   genome.genome_name
316273.25   64.56   Xanthomonas campestris pv. vesicatoria str. 85-10
</pre>
</dd>
<dt>What roles are found in Vibrio campbellii but not Vibrio alginolyticus?</dt>
<dd>To answer this question, we need a file of roles from Vibrio alginolyticus and use it to filter out roles from
Vibrio campbellii. The following pipe gets all the roles from Vibrio alginolyticus genomes and puts them in the file <b>aRoles.tbl</b>.
<pre class="call">
p3-all-genomes --eq "genome_name,Vibrio alginolyticus" | p3-get-genome-features --attr product | p3-function-to-role | p3-sort --count feature.role &gt;aRoles.tbl
</pre>
There are a lot of pieces to this pipe. First, <b>p3-all-genomes</b> gets all the genome IDs for Vibrio alginolyticus. Then <b>p3-get-genome-features</b>
finds all the features for those genomes and outputs the functional assignment (product). <b>p3-function-to-role</b> converts the functions to roles and
eliminates the hypotheticals. Finally, <b>p3-sort</b> with the <code>--count</code> option counts the number of occurrences of each role. It takes a while,
but the output looks something like this.
<pre class="resp">
feature.role    count
(2E,6E)-farnesyl diphosphate synthase (EC 2.5.1.10) 34
(3R)-hydroxymyristoyl-[ACP] dehydratase (EC 4.2.1.-)    34
1,4-alpha-glucan (glycogen) branching enzyme, GH-13-type (EC 2.4.1.18)  37
1,4-alpha-glucan branching enzyme (EC 2.4.1.18) 34
1,4-dihydroxy-2-naphthoate polyprenyltransferase (EC 2.5.1.74)  34
1,4-dihydroxy-2-naphthoyl-CoA hydrolase (EC 3.1.2.28) in menaquinone biosynthesis   34
1,6-anhydro-N-acetylmuramyl-L-alanine amidase   35
1-deoxy-D-xylulose 5-phosphate reductoisomerase (EC 1.1.1.267)  34
1-deoxy-D-xylulose 5-phosphate synthase (EC 2.2.1.7)    35
1-hydroxy-2-methyl-2-(E)-butenyl 4-diphosphate synthase (EC 1.17.7.1)   38
1-phosphofructokinase (EC 2.7.1.56) 34</pre>
Now we perform the same exercise with Vibrio campbellii.
<pre class="call">
p3-all-genomes --eq "genome_name,Vibrio campbellii" | p3-get-genome-features --attr product | p3-function-to-role | p3-sort --count feature.role &gt;cRoles.tbl
</pre>
<pre class="resp">
feature.role    count
(2E,6E)-farnesyl diphosphate synthase (EC 2.5.1.10) 25
1,4-alpha-glucan (glycogen) branching enzyme, GH-13-type (EC 2.4.1.18)  27
1,4-alpha-glucan branching enzyme (EC 2.4.1.18) 25
1,4-dihydroxy-2-naphthoate polyprenyltransferase (EC 2.5.1.74)  27
1,4-dihydroxy-2-naphthoyl-CoA hydrolase (EC 3.1.2.28) in menaquinone biosynthesis   28
1,6-anhydro-N-acetylmuramyl-L-alanine amidase   25
1-deoxy-D-xylulose 5-phosphate reductoisomerase (EC 1.1.1.267)  30
1-deoxy-D-xylulose 5-phosphate synthase (EC 2.2.1.7)    27
1-hydroxy-2-methyl-2-(E)-butenyl 4-diphosphate synthase (EC 1.17.7.1)   26
1-phosphofructokinase (EC 2.7.1.56) 33
16 kDa heat shock protein A 25
</pre>
Now we filter <b>cRoles.tbl</b> by removing records that match <b>aRoles.tbl</b>. Note that we are matching on the key column ONLY. We don't care about the counts,
only which roles are in campbellii but not alginolyticus. The <b>p3-file-filter</b> command performs this task.
<pre class="call">
p3-file-filter --reverse --col=feature.role aRoles.tbl &lt;cRoles.tbl</pre>
The <code>--reverse</code> option tells us we want roles that are in the standard input file (<b>cRoles.tbl</b>) but not the filter file (<b>aRoles.tbl</b>).
The <code>--col</code> option tells us we are comparing values in the <i>feature.role</i> column. Both files are the same format; if the formats were different,
we could specify a different key column identifier for the filter file by appending it as a positional parameter. So, another way to code the same thing
would be
<pre class="call">
p3-file-filter --reverse --col=feature.role aRoles.tbl feature.role &lt;cRoles.tbl
</pre>
The output looks something like this
<pre class="resp">
feature.role    count
2,3-dihydroxybenzoate-AMP ligase (EC 2.7.7.58) of siderophore biosynthesis  33
2-octaprenyl-3-methyl-6-methoxy-1,4-benzoquinol hydroxylase (EC 1.14.13.-)  1
2-pyrone-4,6-dicarboxylic acid hydrolase (EC 3.1.1.57)  14
23S ribosomal RNA rRNA prediction is too short  1
3-polyprenyl-4-hydroxybenzoate carboxy-lyase UbiX (EC 4.1.1.-)  1
4-amino-4-deoxy-L-arabinose transferase 26
4-carboxy-2-hydroxymuconate-6-semialdehyde dehydrogenase    17
4-carboxy-4-hydroxy-2-oxoadipate aldolase (EC 4.1.3.17) 16
4-hydroxy-2-oxovalerate aldolase (EC 4.1.3.39)  7
4-oxalmesaconate hydratase (EC 4.2.1.83)    14
4-oxalocrotonate tautomerase    1
</pre>
</dd>
<dt>What is the average length of proteins in family PGF_00112374?</dt>
<dd>If we had a file of protein family names with the amino acid length of each protein in the family, we can use the script
<b>p3-stats</b> to output the mean length as well as the minimum, count, maximum, and standard deviation. The following pipe
does the trick.
<pre class="call">
p3-echo -t family PGF_00112374 | p3-get-family-features --ftype=global --attr aa_length | p3-stats --col=family feature.aa_length
</pre>
<pre class="resp">
family  count   average min max stdev
PGF_00112374    3414    818.125659050967    31  901 193.491091039707
</pre>
The <b>p3-echo</b> command creates a one-line file with the family ID in it. We use <b>p3-get-family-features</b> to get all the
features in this family. The <code>--ftype=global</code> parameter indicates that this is a global protein family (there are also
families of type <i>local</i> and <i>figfam</i>). For each feature, we want the amino acid length. This value is stored in the
<i>aa_length</i> attribute. Finally, we have <b>p3-stats</b>. The <code>--col=family</code> parameter tells us the input file
records are to be grouped by the content of the <i>family</i> column. The positional <code>feature.aa_length</code> parameter
tells us the numbers to analyze can be found in the <i>aa_length</i> column from the <i>feature</i> record. The output tells us
there are 3414 pegs in the family. The average length is a little over 818 amino acids with a standard deviation of well over 193.
The total range is 31 amino acids to 901 amino acids.
</dd>
<dt>Display the CDS and RNA features for genome 1313.7001 sorted by location on the chromosome.</dt>
<dd>
<pre class="call">
p3-echo -t genome_id 1313.7001 | p3-get-genome-features --in feature_type,CDS,rna --attr patric_id --attr sequence_id --attr start --attr strand --attr product | p3-sort feature.sequence_id feature.start/n feature.strand
</pre>
We start by using <b>p3-echo</b> to create a file that has our single genome ID in it. The bulk of the retrieval work is
performed by <b>p3-get-genome-features</b>. The <code>--in</code> parameter allows us to specify a list of values for a
specific field. In this case, we want <i>feature_type</i> to equal either <code>CDS</code> or <code>rna</code>.
To sort by location, we need the contig ID (<i>sequence_id</i>) and the start location (<i>start</i>). The start location
is always the leftmost location on the contig, so it is perfect for sorting. Finally, we add the strand (<code>+</code> or <code>i</code>)
and then the functional assignment (<i>product</i>) so we can see what the feature does. The <b>p3-sort</b> gets the
file records in the proper order. Because one of the fields is numeric, we put a <code>/n</code> after the field name.
This tells the sorter that for the <i>feature.start</i> column, the value <code>20</code> comes before, not after, the
value <code>100</code>. The output will look something like this.

<pre class="resp">
genome_id   feature.patric_id   feature.sequence_id feature.start   feature.strand  feature.product
1313.7001   fig|1313.7001.peg.1 1313.7001.con.0001  40  -   Mobile element protein
1313.7001   fig|1313.7001.peg.2 1313.7001.con.0001  540 -   Mobile element protein
1313.7001   fig|1313.7001.peg.3 1313.7001.con.0001  994 -   Mobile element protein
1313.7001   fig|1313.7001.peg.4 1313.7001.con.0002  1   +   Streptococcal histidine triad protein
1313.7001   fig|1313.7001.peg.5 1313.7001.con.0003  1   +   Endo-beta-N-acetylglucosaminidase (EC 3.2.1.96)
1313.7001   fig|1313.7001.peg.6 1313.7001.con.0003  1120    -   Fibronectin/fibrinogen-binding protein
1313.7001   fig|1313.7001.peg.7 1313.7001.con.0003  2880    +   Metal-dependent hydrolase YbeY, involved in rRNA and/or ribosome maturation and assembly
1313.7001   fig|1313.7001.peg.8 1313.7001.con.0003  3358    +   Diacylglycerol kinase (EC 2.7.1.107)
1313.7001   fig|1313.7001.peg.9 1313.7001.con.0003  3770    +   GTP-binding protein Era
1313.7001   fig|1313.7001.peg.10    1313.7001.con.0003  4684    +   Formamidopyrimidine-DNA glycosylase (EC 3.2.2.23)
1313.7001   fig|1313.7001.peg.11    1313.7001.con.0003  5541    +   Dephospho-CoA kinase (EC 2.7.1.24)
1313.7001   fig|1313.7001.peg.12    1313.7001.con.0003  6133    +   Multidrug resistance efflux pump PmrA
1313.7001   fig|1313.7001.peg.13    1313.7001.con.0003  7521    +   Protein translocase membrane subunit SecG
1313.7001   fig|1313.7001.peg.14    1313.7001.con.0003  7856    +   3'-to-5' exoribonuclease RNase R
1313.7001   fig|1313.7001.peg.15    1313.7001.con.0003  10173   +   tmRNA-binding protein SmpB
1313.7001   fig|1313.7001.peg.16    1313.7001.con.0003  10656   +   Tellurite methyltransferase (EC 2.1.1.265)
</pre>
</dd>
<dt>Compute the upstream regions for the protein-encoding genes in genome 1313.7001.</dt>
<dd>We get upstream regions from the <b>p3-feature-upstream</b> script, but to use it we need an input list of feature IDs.
We will produce feature IDs and functional assignments, then append the upstream sequences.
<pre class="call">
p3-echo -t genome_id 1313.7001 | p3-get-genome-features --eq feature_type,CDS --attr patric_id --attr product | p3-feature-upstream --col=feature.patric_id
</pre>
The <code>-eq feature_type,CDS</code> ensures we only see protein-encoding features. Because we are not putting the feature ID in the last column,
we use <code>--col=feature.patric_id</code> to direct <b>p3-feature-upstream</b> to the correct input column. The output will look something like this.
Note the upstream DNA is in the last column.
<pre class="resp">
genome_id   feature.patric_id   feature.product upstream
1313.7001   fig|1313.7001.peg.1182  beta-glycosyl hydrolase      ttgtcatctcctcttgactctcgttaatataagaaataaaataagggcgttgatttatataatcgctatcaatataacaatgcaatcaggaggttttgca
1313.7001   fig|1313.7001.peg.1189  IMP cyclohydrolase (EC 3.5.4.10) / Phosphoribosylaminoimidazolecarboxamide formyltransferase (EC 2.1.2.3)   gatcaatatcttaggtatgcttagccttggttttgcttatcttgttttactgttactgcatttaattggtgtttaactaatgattaaaaaggagaatata
1313.7001   fig|1313.7001.peg.1191  Phosphoribosylglycinamide formyltransferase (EC 2.1.2.2)      tcagccctgaaaatgtagagcgtgtaaaagaattgttggatgaagcagtctatgaaattggtcgcatcgtcaagaaagaaaacgaaagtgtcattatcaa
1313.7001   fig|1313.7001.peg.1192  Phosphoribosylformylglycinamidine cyclo-ligase (EC 6.3.3.1)   tctctatgactacgaagaagactatcgtagaagtttggaagaaaagaccagtttttacaagtaggcgacagattctccattaaagaaaaggaaaaaacaa
1313.7001   fig|1313.7001.peg.1199  hypothetical protein    aaggtggcggatgcaattggggagattttgccaaagcaggtgttggaggaggagctatacttggaggtgtggcctatgcagcgacatgttggtggtaatt
1313.7001   fig|1313.7001.peg.1211  hypothetical protein    ttggcgattaccaacaatggacaggaaaaccatctggttaagatggcattcttggaattaaaaaatacagagaaaccagcaaagacaaggttcgcaagcc
1313.7001   fig|1313.7001.peg.1259  Acetyl xylan esterase 1; Cephalosporin-C deacetylase (EC 3.1.1.41)  aaagaatctaaattcactttctatttacccttctttcttgcattgattacatagatatgctacagttgtggtaacgattacaaaataaaaggagcatgct
1313.7001   fig|1313.7001.peg.1278  Helicase loader DnaB    acgttttgctagtgtctatcgtagttttaaggatgtcagtgagttagagagcttgctccaacaaatcacccagtcctctaaaaagaaaaaggaaagataa
1313.7001   fig|1313.7001.peg.1288  Fructokinase (EC 2.7.1.4)   ttattagatagtaagatttacagaggaaaatctaaaaaatagagacatttagactttcgaagtatgctataataaagaaaataaaaacaagaggtttatc
</pre>
You can use the <code>--len</code> parameter of <b>p3-feature-upstream</b> to change the number of base pairs displayed (the default is 100). If the feature is
at the edge of the contig, you may see less than the specified length or even nothing at all, since the script stops at the contig boundary. To see downstream
regions instead, use the <code>--downstream</code> option. This pipe shows the 10 base pairs downstream of each gene.
<pre class="call">
p3-echo -t genome_id 1313.7001 | p3-get-genome-features --eq feature_type,CDS --attr patric_id --attr product | p3-feature-upstream --col=feature.patric_id --downstream --len=10
</pre>
<pre class="resp">
genome_id   feature.patric_id   feature.product downstream
1313.7001   fig|1313.7001.peg.1182  beta-glycosyl hydrolase    gtcttttcga
1313.7001   fig|1313.7001.peg.1189  IMP cyclohydrolase (EC 3.5.4.10) / Phosphoribosylaminoimidazolecarboxamide formyltransferase (EC 2.1.2.3)   gaagataaaa
1313.7001   fig|1313.7001.peg.1191  Phosphoribosylglycinamide formyltransferase (EC 2.1.2.2)    ctttttgatg
1313.7001   fig|1313.7001.peg.1192  Phosphoribosylformylglycinamidine cyclo-ligase (EC 6.3.3.1)     aaaaaatagc
1313.7001   fig|1313.7001.peg.1199  hypothetical protein    tcaaaactat
1313.7001   fig|1313.7001.peg.1211  hypothetical protein    tcaactacat
1313.7001   fig|1313.7001.peg.1259  Acetyl xylan esterase 1; Cephalosporin-C deacetylase (EC 3.1.1.41)  ggagtcgact
1313.7001   fig|1313.7001.peg.1278  Helicase loader DnaB    atggaaagtg
</pre>
</dd>
<dt>What is the codon usage in genome 186497.12?</dt>
<dd>The <b>p3-sequence-profile</b> script counts the number of occurrences of each letter in a sequence field.
To use it, we need to create a file that has the sequences we want to analyze in the last column.
We start with the genome ID, then use <b>p3-get-genome-features</b> to get the feature data. The <i>aa_sequence</i>
field contains the protein sequences, which are then processed by <b>p3-sequence-profile</b>.
<pre class="call">
p3-echo -t genome_id 186497.12 | p3-get-genome-features --attr aa_sequence | p3-sequence-profile
</pre>
By default, <b>p3-sequence-profile</b> works on the last input column, which in this case is the amino acid
sequence. The output will look something like this.
<pre class="resp">
letter  count
L   58114
E   51852
I   50270
K   46874
V   45417
G   41210
A   38057
R   30791
S   28102
F   25399
T   25375
D   25340
P   24706
Y   23048
N   19998
M   12966
Q   10045
H   8653
W   7104
C   3359
</pre>
Note that the output is sorted from most common to least. The same trick works for DNA sequences, which are in the
<i>na_sequence</i> field.
<pre class="call">
p3-echo -t genome_id 186497.12 | p3-get-genome-features --attr na_sequence | p3-sequence-profile
</pre>
<pre class="resp">
letter  count
A   597286
T   465185
G   440197
C   305118
</pre>
</dd>
<dt>How do I find a genome from an accession number or project ID?</dt>
<dd>Genomes in PATRIC are stored with four alternate IDs, any of which can be used to search for the genomes.
<ol>
<li>ncbi_project_id. <span style="font-weight: normal">The NCBI project number</span></li>
<li>refseq_project_id. <span style="font-weight: normal">The REFSEQ project number</span></li>
<li>genbank_accessions. <span style="font-weight: normal">The accession string from GENBANK</span></li>
<li>refseq_accessions. <span style="font-weight: normal">The accession string from REFSEQ</span></li>
</ol>
The following commands all return the genome <i>Streptococcus mutans UA159</i>.
<pre class="call">
p3-all-genomes --eq genbank_accessions,AE014133
</pre>
<pre class="call">
p3-all-genomes --eq refseq_accessions,NC_004350
</pre>
<pre class="call">
p3-all-genomes --eq ncbi_project_id,333
</pre>
<pre class="call">
p3-all-genomes --eq refseq_project_id,57947
</pre>
Because no attributes were specified, the output in each case is solely the genome ID, a single output record in a single column.
<pre class="resp">
genome.genome_id
210007.7
</pre>
</dd>
<dt>How do I find a feature from an alternate feature ID?</dt>
<dd>Features in PATRIC are stored with four alternate IDs, all of which are indexed for fast retrieval.
<ol>
<li>gene. <span style="font-weight: normal">The common gene name (e.g. <code>ciaR</code>). Not all features will have a common gene ID</span></li>
<li>gene_id. <span style="font-weight: normal">The gene number</span></li>
<li>gi. <span style="font-weight: normal">The GI accession number</span></li>
<li>refseq_locus_tag. <span style="font-weight: normal">The locus tag from REFSEQ</span></li>
</ol>
The command <b>p3-find-features</b> is used to retrieve features based on an alternate ID. The alternate IDs are piped in through the standard input.
Note that in some cases (e.g. the field <code>gene</code>), there may be a lot of features with the same ID. The following pipes return the ID and
functional assignment using alternate IDs.
<pre class="call">
p3-echo coaA | p3-find-features --attr patric_id,product gene
</pre>
<pre class="resp">
id  feature.patric_id   feature.product
coaA    fig|996634.5.peg.916    Pantothenate kinase (EC 2.7.1.33)
coaA    fig|944560.4.peg.377    Pantothenate kinase (EC 2.7.1.33)
coaA    fig|992133.3.peg.4201   Pantothenate kinase (EC 2.7.1.33)
coaA    fig|992141.3.peg.4166   Pantothenate kinase (EC 2.7.1.33)
coaA    fig|992132.3.peg.4176   Pantothenate kinase (EC 2.7.1.33)
coaA    fig|992136.3.peg.3951   Pantothenate kinase (EC 2.7.1.33)
coaA    fig|992139.3.peg.4173   Pantothenate kinase (EC 2.7.1.33)
coaA    fig|992131.3.peg.3905   Pantothenate kinase (EC 2.7.1.33)
coaA    fig|992142.3.peg.3915   Pantothenate kinase (EC 2.7.1.33)
coaA    fig|992137.3.peg.4122   Pantothenate kinase (EC 2.7.1.33)
coaA    fig|992135.3.peg.3895   Pantothenate kinase (EC 2.7.1.33)
coaA    fig|99287.12.peg.4355   Pantothenate kinase (EC 2.7.1.33)
coaA    fig|996306.3.peg.645    Pantothenate kinase (EC 2.7.1.33)
coaA    fig|992175.3.peg.3998   Pantothenate kinase (EC 2.7.1.33)
coaA    fig|992179.3.peg.3888   Pantothenate kinase (EC 2.7.1.33)
coaA    fig|992177.3.peg.4127   Pantothenate kinase (EC 2.7.1.33)
coaA    fig|992178.3.peg.4122   Pantothenate kinase (EC 2.7.1.33)
coaA    fig|992180.3.peg.4094   Pantothenate kinase (EC 2.7.1.33)
coaA    fig|992172.3.peg.3873   Pantothenate kinase (EC 2.7.1.33)
coaA    fig|992181.3.peg.3864   Pantothenate kinase (EC 2.7.1.33)
</pre>
<pre class="call">
p3-echo 1029377 | p3-find-features --attr patric_id,product gene_id
</pre>
<pre class="resp">
id  feature.patric_id   feature.product
1029377 fig|210007.7.peg.1009   Pantothenate kinase (EC 2.7.1.33)
</pre>
<pre class="call">
p3-echo 24379558 | p3-find-features --attr patric_id,product gi
</pre>
<pre class="resp">
id  feature.patric_id   feature.product
24379558    fig|210007.7.peg.1009   Pantothenate kinase (EC 2.7.1.33)
</pre>
<pre class="call">
p3-echo SMU.1126 | p3-find-features --attr patric_id,product refseq_locus_tag
</pre>
<pre class="resp">
id  feature.patric_id   feature.product
SMU.1126    fig|210007.7.peg.1009   Pantothenate kinase (EC 2.7.1.33)
</pre>
</dd>
<dt>Given a genome ID, how do I find the name?</dt>
<dd>The genome name is in an attribute called <i>genome_name</i>. You can get it from the genome ID using <b>p3-all-genomes</b> as shown here.
<pre class="call">
p3-all-genomes --eq genome_id,210007.7 --attr genome_name
</pre>
<pre class="resp">
genome.genome_id    genome.genome_name
210007.7    Streptococcus mutans UA159
</pre>
Alternatively, you can use <b>p3-get-genome-data</b> and use <b>p3-echo</b> to pipe in the ID.
<pre class="call">
p3-echo 210007.7 | p3-get-genome-data --attr genome_name
</pre>
<pre class="resp">
id  genome.genome_name
210007.7    Streptococcus mutans UA159
</pre>
</dd>
<dt>Given a feature ID, how do I find the amino acid sequence?</dt>
<dd>The amino acid sequence is in the attribute <i>aa_sequence</i>. You use <b>p3-get-feature-data</b> to access it.
<pre class="call">
p3-echo "fig|210007.7.peg.1009" | p3-get-feature-data --attr aa_sequence
</pre>
<pre class="resp">
id  feature.aa_sequence
fig|210007.7.peg.1009   MANEFINFEKISRKTWQHLHQESQPPLNENELNSIKSLNDRISIKDVTDIYLPLISLIQIYKKSQENLSFSKSIFLQKNISNRPFIIGVSGSVAVGKSTTSRLLQLLLARTFKDSSVELMTTDGFLYPNAVLSSRHMLNKKGFPESYDMERLLDFLDTIKNGQSAEIPVYSHEIYDIVPNKSQIIEVPDFLIIEGINVFQNPQNNRLYMSDFFDFSIYIDADSDYIENWYLERFATLLDLAKNDKQNYYNRFLKLGEKGALDFARDIWKDINLVNLEKYIEPTRSRAELILHKTKNHKIDEIYLKK
</pre>
</dd>
<dt>How do I find out how many genomes have an identical protein to a given feature?</dt>
<dd>The individual protein sequences are not indexed, but the PATRIC database contains an MD5 signature for each protein that is indexed, in the feature
attribute <i>aa_sequence_md5</i>. The following pipe finds out how many times the protein sequence for <i>fig|210007.7.peg.1009</i> occurs in the database.
<pre class="call">
p3-echo "fig|210007.7.peg.1009" | p3-get-feature-data --attr aa_sequence_md5 | p3-find-features aa_sequence_md5 --count
</pre>
<pre class="resp">
id      feature.aa_sequence_md5 feature.count
fig|210007.7.peg.1009   6400069a6f7f32515c3a584ade0588d0        150
</pre>
The answer is 150, but that is not quite the question that was asked. If the protein occurs more than once in a genome, then the above count
will be too high. To get the correct answer we need to extract genome IDs and then count the number of distinct ones with <b>p3-count</b>.
<pre class="call">
p3-echo "fig|210007.7.peg.1009" | p3-get-feature-data --attr aa_sequence_md5 | p3-find-features aa_sequence_md5 --attr genome_id | p3-count genome_id
</pre>
<pre class="resp">
count
149
</pre>
</dd>
<dt>How do I find all the genomes in a given genus?</dt>
<dd>The <b>p3-all-genomes</b> script is used to do this. Simply filter on the <i>genus</i> attribute.
<pre class="call">
p3-all-genomes --eq genus,Streptococcus --attr genome_id,genome_name
</pre>
<pre class="resp">
genome.genome_id    genome.genome_name
1302.21 Streptococcus gordonii strain DD07
1303.76 Streptococcus oralis strain DD05
1303.77 Streptococcus oralis strain DD14
1303.78 Streptococcus oralis strain DD15
1303.79 Streptococcus oralis strain DD16
1303.80 Streptococcus oralis strain DD20
1303.81 Streptococcus oralis strain DD21
1303.82 Streptococcus oralis strain DD27
1303.83 Streptococcus oralis strain DD30
1303.84 Streptococcus oralis strain DD24
1303.86 Streptococcus oralis strain DD25
1303.87 Streptococcus oralis strain DD17
1318.28 Streptococcus parasanguinis strain DD19
28037.233   Streptococcus mitis strain DD26
28037.234   Streptococcus mitis strain DD28
28037.238   Streptococcus mitis strain DD22
</pre>
</dd>
<dt>Given a feature ID, how do I find the features in the same protein family?</dt>
<dd>The family ID is in the <i>pgfam_id</i> attribute, and we use <b>p3-get-family-features</b> with the <code>--ftype=global</code> to find the other features.
<pre class="call">
p3-echo "fig|210007.7.peg.1009" | p3-get-feature-data --attr pgfam_id | p3-get-family-features --ne "patric_id,fig|210007.7" --ftype global --attr patric_id,genome_name,product
</pre>
Note that we use the <code>--ne</code> operator to keep the original feature from appearing in the output. Even so, the resulting file has over 67,000 results, the first
few of which are shown below.
<pre class="resp">
id  feature.pgfam_id    feature.patric_id   feature.genome_name feature.product
fig|210007.7.peg.1009   PGF_00029921    fig|1341640.3.peg.2368  Yersinia sp. WP-930601  Pantothenate kinase (EC 2.7.1.33)
fig|210007.7.peg.1009   PGF_00029921    fig|1341642.3.peg.3368  Yersinia sp. WP-931205  Pantothenate kinase (EC 2.7.1.33)
fig|210007.7.peg.1009   PGF_00029921    fig|1344012.3.peg.2066  Tatumella sp. NML 06-3099   Pantothenate kinase (EC 2.7.1.33)
fig|210007.7.peg.1009   PGF_00029921    fig|984229.3.peg.2006   Salmonella enterica subsp. enterica serovar Enteritidis str. 653049 13-19   Pantothenate kinase (EC 2.7.1.33)
fig|210007.7.peg.1009   PGF_00029921    fig|984228.3.peg.4305   Salmonella enterica subsp. enterica serovar Enteritidis str. 648905 5-18    Pantothenate kinase (EC 2.7.1.33)
fig|210007.7.peg.1009   PGF_00029921    fig|984226.3.peg.4289   Salmonella enterica subsp. enterica serovar Enteritidis str. 648903 1-6 Pantothenate kinase (EC 2.7.1.33)
fig|210007.7.peg.1009   PGF_00029921    fig|984227.3.peg.3102   Salmonella enterica subsp. enterica serovar Enteritidis str. 648904 3-6 Pantothenate kinase (EC 2.7.1.33)
fig|210007.7.peg.1009   PGF_00029921    fig|984224.3.peg.642    Salmonella enterica subsp. enterica serovar Enteritidis str. 648901 39-2    Pantothenate kinase (EC 2.7.1.33)
fig|210007.7.peg.1009   PGF_00029921    fig|984225.3.peg.4221   Salmonella enterica subsp. enterica serovar Enteritidis str. 648902 6-8 Pantothenate kinase (EC 2.7.1.33)
</pre>
Alternatively, you can use local protein families (<i>plfam_id</i> for the field name, and <code>--ftype=local</code> for the family type). This will restrict the output
to features for genomes in the same genus.
<pre class="call">
p3-echo "fig|210007.7.peg.1009" | p3-get-feature-data --attr plfam_id | p3-get-family-features --ne "patric_id,fig|210007.7" --ftype local --attr patric_id,genome_name,product
</pre>
<pre class="resp">
id  feature.plfam_id    feature.patric_id   feature.genome_name feature.product
fig|210007.7.peg.1009   PLF_1301_00006228   fig|1579339.3.peg.1420  Streptococcus sp. 449_SSPC  Pantothenate kinase (EC 2.7.1.33)
fig|210007.7.peg.1009   PLF_1301_00006228   fig|511691.3.peg.905    Streptococcus mutans NN2025 Pantothenate kinase (EC 2.7.1.33)
fig|210007.7.peg.1009   PLF_1301_00006228   fig|1404260.3.peg.1014  Streptococcus mutans PKUSS-LG01 Pantothenate kinase (EC 2.7.1.33)
fig|210007.7.peg.1009   PLF_1301_00006228   fig|1403829.3.peg.1035  Streptococcus mutans PKUSS-HG01 Pantothenate kinase (EC 2.7.1.33)
fig|210007.7.peg.1009   PLF_1301_00006228   fig|857136.3.peg.247    Streptococcus mutans A19    Pantothenate kinase (EC 2.7.1.33)
fig|210007.7.peg.1009   PLF_1301_00006228   fig|857135.3.peg.392    Streptococcus mutans U138   Pantothenate kinase (EC 2.7.1.33)
fig|210007.7.peg.1009   PLF_1301_00006228   fig|857134.3.peg.201    Streptococcus mutans G123   Pantothenate kinase (EC 2.7.1.33)
fig|210007.7.peg.1009   PLF_1301_00006228   fig|857133.3.peg.623    Streptococcus mutans M21    Pantothenate kinase (EC 2.7.1.33)
fig|210007.7.peg.1009   PLF_1301_00006228   fig|857132.3.peg.561    Streptococcus mutans T4 Pantothenate kinase (EC 2.7.1.33)
fig|210007.7.peg.1009   PLF_1301_00006228   fig|857138.3.peg.734    Streptococcus mutans N29    Pantothenate kinase (EC 2.7.1.33)
fig|210007.7.peg.1009   PLF_1301_00006228   fig|857137.3.peg.82 Streptococcus mutans NMT4863    Pantothenate kinase (EC 2.7.1.33)
fig|210007.7.peg.1009   PLF_1301_00006228   fig|1313.8640.peg.2062  Streptococcus pneumoniae strain B16827  Pantothenate kinase (EC 2.7.1.33)
</pre>
</dd>
<dt>How do I find a gene by name in a particular genome?</dt>
<dd>Here you want to use <b>p3-find-features</b> with a genome_id filter.
<pre class="call">
p3-echo coaA | p3-find-features --attr patric_id,product --eq genome_id,210007.7 gene
</pre>
<pre class="resp">
id  feature.patric_id   feature.product
coaA    fig|210007.7.peg.1009   Pantothenate kinase (EC 2.7.1.33)
</pre>
</dl>
<p>&nbsp;</p>
</div>
</body>
</html>
