<!doctype html>
<html>
<head>
<meta charset="utf-8" />
<title>The Representative-Genomes Server</title>
<link href="../css/Basic.css" rel="stylesheet" type="text/css">
</head>
<body>
<div id="Dump">
<h1>The Representative-Genomes Server</h1>

<p>In early 2017, as I write this document, I am pondering the following
three points:</p>
<ol>
<li>PATRIC now contains about 90,000 genomes.</li>
<li>Approximately 7500 of these genomes are marked as
complete.  We are running evaluations of all of the
PATRIC genomes as I write this, and I believe that when
we complete the efort there will be a close
correspondence between the genomes marked as <i>complete</i>
and those marked as <i>good</i> by the evaluation process
(essentially, a <i>good genome</i> has a checkM completeness
score of at least 80 and an FIGannCon score of at least
85).</li>
<li>The majority of the 7500 complete/good genomes represent
pathogens (with genomes that are very similar).  You can
pick about 2000 representative genomes from the 7500.</li>
</ol>


<p>While there is great need for sets of very similar genomes to support
SNP analysis and studies of AMR, there is also a need for the
representative genomes to support comparative analysis of more
diverged genomes.  From my perspective, these two broad types of
comparative analysis (one of very similar genomes, and one of diverse
genomes that cover the tree) both need to be supported by PATRIC.</p>

<p>My personal interests center on issues that require a broad set of
genomes that cover the tree.  We are working with Rob Edwards and his
group to gather a large, diverse set of good genomes and make them
freely available.</p>

<p>The main sources of good genomes at this point are genomes imported
from</p>

<ol>
<li>the CoreSEED [the manually curated collection being
maintained as part of the SEED Project]</li>
<li>the good genomes from PATRIC</li>
<li>good genomes harvested from a rapidly growing body of
metagenomic samples.</li>
</ol>


<p>These three sources are being merged into two collections:</p>

<ol>
<li>The first collection is composed of the representative
genomes and will be maintained within the SEEDtk
environment. Frequent annotation updates from the CoreSEED
effort are critical to maintaing the value of this
collection (biweekly updates are currently taking place).
</li>
<li>The second collection is a set of genome sets, each
representing a collection of very similar genomes.</li>
</ol>

<h2>Similarity</h2>

<p>I have used the concept of <i>similar genomes</i> without making the notion
precise.  Since we are entering a period in which millions of genomes
will be sequenced, the question of how to construct tools that scale
quickly emerges.  In particular, we need a simple notion of the
critical concept of <i>similarity of two genomes</i>.

<p>The functional role <b>Phenylalanyl-tRNA synthetase alpha chain (EC 6.1.1.20)</b>

<p>is almost universal (this is the gene product of pheS in Escherichia
coli).  Very occasionally it might be part of a fusion or a
frameshift, but for our purposes, we will constrain the notion of
<i>good genome</i> to require that a good genome includes precisely one
copy of pheS and the encoded protein will have a length that falls
within two standard deviations of the mean (the mean length is about
347 amino acids and the standard deviation is about 46 aa).</p>

<p>Now we can define the similarity between two good genomes as the
number of amino acid 8-mers that are in common between the gene
products of the two PheS genes.</p>

<h2>Computation Steps Needed to Support Representative Genomes</h2>

<h3>Step 1: Computation of Very Similar Sets (VSS)</h3>

<p>I will now go through the steps needed to support computation of
representative genomes.  A user begins by creating two files:</p>

<dl>
<dt>complete.genomes</dt>
<dd>a 2-column text file. Each row contains [Genome ID, Genome name].</dd>
<dt>6.1.1.20.fasta</dt>
<dd>a fasta file of the amino acid sequences of the PheS products for
the set of genomes in the <b>complete.genomes</b> file.</dd>
</dl>

<p>Then use something like</p>

<pre class="call">
mkdir Data
cp complete.genomes Data
cp 6.1.1.20.fasta Data
get_vss_for_reps Data
</pre>

<p>The <b>get_vss_for_reps</b> command computes the file <i>Data/vss</i>, which contains sets of very
similar sequences.</p>

<h3>Step 2: Computing the Distance Matrix</h3>

<p>It is now necessary to specify the kmer size. This is accomplished by placing
a number (of amino acids per kmer) in a file named <b>K</b> in the <b>Data</b> directory.</p>

<pre class="call">
echo 8 >Data/K
</pre>

<p>Now</p>

<pre class="call">
generate_cached_rep_data Data
</pre>

<p>generates a <i>similarity matrix</i> for the genomes in <b>Data/complete.genomes</b> and stores it in <b>Data/similarities</b>.
It also creates <b>Data/genome.index</b> which creates a surrogate key for each of the genomes that
will be used in computing representative sets.</p>

<p>Note that, at this time, we have a very tractable set of genomes,
but when we reach 100,000 divergent genomes things may get worse.  The
hope is that we will be covered for 5-10 years (at which time
computers will be free).</p>

<p>When I designed this tool initially, I thought that there would be a
need for high-performance, and this lead to the creation of surrogate
keys for genomes, where the surrogates are consecutive integers
(allowing tables to be built using integers -- speeding some
computations and reducing memory requirements.  I foresaw a time where
the core computations would be pushed into C.</p>

<h3>Step 3: The Server</h3>

<p>Eventually, I think we will need a server that</p>

<ol>
<li>reads in precomputed values (most notably Data/similarities) and</li>
<li>processes a stream of requests.</li>
</ol>

<p>To make the construction of the server easier, I have implemented a
query-processor similar to the old <b>fig.pl</b> style.  I debug and time
implementations of requests in that environment.  The query processor
is invoked using</p>

<pre class="call">
rep_server Data
</pre>

<p>After loading the precomputed data, a prompt should appear (this can
take a few seconds to appear).  A request of just <code>h</code> produces a short
help text.</p>

<p>Here is an example interaction:</p>

<p>This invokes the query processor</p>
<pre class="call">
$ rep_server Data
</pre>
<p>This asks for the genome id and name, given an index value of 2</p>
<pre class="call">
?? index_to_id 2
        2   1680.5  Bifidobacterium adolescentis strain BBMN23
        3.60012054443359e-05 seconds to execute command
</pre>
<p>Note that the request is timed.</p>

<p>This produces the same information, given the id of the genome.</p>
<pre>
?? id_to_index 1680.5
        2   1680.5  Bifidobacterium adolescentis strain BBMN23
        3.60012054443359e-05 seconds to execute command
</pre>
<p>This requests the help text be displayed</p>
<pre class="call">
?? h

            closest_genomes Index N    [returns closest N genomes]
            close_rep_seq N Sequence   [returns close representative]
            id_to_index  Id            [returns genome index]
            index_to_id  Index         [returns genome id]
            match_tails 16-mer         [returns genomes with matching PheS tail]
            n_reps N [keep1, keep2, ...keepn]
                                       [returns rep set of about N seqs]
            rep_set N [keep1, keep2, ...keepN]
                                       [ returns rep set ]
            thin_set N Index1 Index2 ... IndexN  [ make thinned set ]
        5.19752502441406e-05 seconds to execute command
</pre>
<p>Finally, we exit</p>
<pre class="call">
?? x
$
</pre>

<h4>Other commands</h4>

<p>Assuming that you have a genome from the set of complete.genomes,
you can use something like</p>
<pre class="call">
?? closest_genomes 2 5
        2   1680.5  Bifidobacterium adolescentis strain BBMN23
                343 16  1680.4  Bifidobacterium adolescentis strain 22L
                343 829 367928.6    Bifidobacterium adolescentis ATCC 15703
                280 1   1447716.4   Bifidobacterium kashiwanohense PV20-2
                280 864 1150460.7   Bifidobacterium kashiwanohense JCM 15439 = DSM 21854
                280 2022    547043.19   Bifidobacterium pseudocatenulatum DSM 20438 = JCM 1200 = LMG 10505
        0.000138998031616211 seconds to execute command
</pre>
<p>to get the closest 5 genomes (i.e., the closest 5 to the genome with
index 2).</p>

<p>You can compute a representative set of genomes.  The algorithm takes
a similarity threshhold N.  The returned set will have the property
that no two members are more similar than N, and all of the omitted
sequences are similar to one of the representative sequences at a
score of N or more.  You can optionally specify a set of <i>keepers</i>, which are indexes of
sequences you want to be present in the representative set (if
possible).  Thus, the following command creates a representative set
based on similarity threshhold of 50, and forces one of the members to
be the genome with index 864. The returned set has 872 members.</p>
<pre class="call">
?? rep_set 50 864
        872

        864 1150460.7   Bifidobacterium kashiwanohense JCM 15439 = DSM 21854
        3   1678678.3   Burkholderia sp. HB1
        5   1532555.3   Brevundimonas sp. DS20
        6   1746199.3   Bordetella sp. N
        10  224911.44   Bradyrhizobium diazoefficiens USDA 110
        11  98804.3 Buchnera aphidicola (Tuberolachnus salignus)
        12  118110.3    Buchnera aphidicola (Schlechtendalia chinensis)
        13  1263829.3   Bibersteinia trehalosi USDA-ARS-USMARC-188

        .
        .
        .
        0.0565729141235352 seconds to execute command
</pre>

<h4>Getting a Representative set with a Specified Number of Members</h4>

<p>You can compute a representative set containing close to a specified
number of members using something like</p>
<pre class="call">
?? n_reps 1000
        1001    0,3,5,6,10,11,12,13,14,15,17,18,19,20,...

        0.0629708766937256 seconds to execute command

        ??
</pre>
<p>Note that the genomes are given as a comma-separated set of genome indices.</p>




<h4>Trying to Find a Close representative sequence</h4>


<p>Suppose that you had an arbitrary PheS sequence, and you wanted to
locate a genome with a close sequence.  You could use something like</p>

<pre class="call">
?? close_rep_seq 200 MGSMDLDQIVADAQQSFEGAADITTLENEKARFLGKSGALTELLKGLGKLDPETRKTEGARINVAKQQVEAALNARRQALADALLNQRLAAEAIDVTLPGRGAGAGSLHPVMRTWERVEQIFRSIGFDVADGPEIETDWYNFTALNSPENHPARSMQDTFYVDGKDADGRPLLLRTHTSPMQVRYARMNRPPIKVIAPGRTYRVDSDATHSPMFNQVEGLWIDENVSFADLKGVYTDFLKKFFERDDILVRFRPSYFPFTEPSAEIDMMFEHGKNAGKWLEISGSGQVHPTVIRNMGLDPERYIGFAFGSGLERLTMLRYGVQDLRLFFENDLRFLRQFA
        268 1779    1009846.3   Burkholderia cepacia GG4
        0.503750085830688 seconds to execute command
</pre>

<p>which tries to find a representative sequence that is close to the
specified sequence.  The search computes a set of about 200 rep sets
that cover the collection, keeping the closest to the given sequence.
It then takes that rep sequence and looks at the 50 closest other rep
sequences.  The closest is returned.  In this case it was 1009846.3
(index=1779) that had 268 8-mers in common with the query PheS.</p>


<h4>Extending a Pin to Support Compare Regions</h4>

<p>Computing a <i>pin</i> is a key step in a compare regions computation.
My intent is to create a strategy that is not guaranteed to produce
the pin, but is fast enough to allow fallback approaches when it
fails.</p>

<p>The basic approach that I advocate would be</p>

<ol>
<li>Get the PheS from the focus genome.  If the focus genome is
in complete.genomes, this PheS has instantaneous access to the
closest genomes from complete.genomes (via <code>closest_genomes</code>
query).  Otherwise, you need to use a <code>close_rep_seq</code> and a
<code>closest_genomes</code> to get the set of close genomes.</li>
<li>Once you have the close genomes, You need to get the set of
pinned features.  If the seed feature is in a protein
family, compute the pin as the set of features from the family
that are from the close genomes.  This can lead to multiple
pegs from the same genome.  I think that is ok, if the seed
feature occurs uniquely (and maybe ok even if it is not unique).</li>
<li>If you cannot easily get a pin from the seed feature, find the
closest non-hypo feature for which you can get a pin, and use
that (adjusting the red to reflect the original seed feature
-- not the close non-hypo).</li>
</ol>
<p>This approach will, in the worst case, take under a second to fail.
If the seed feature is from a genome in <b>complete.genomes</b>, it should
take under 0.1 seconds to fail. My belief is that it will succeed most of the time.</p>

<h2>Identifying Occurrences of the Representative Genomes</h2>

<p>Let us summarize some of the central ideas so far:</p>

<p>We have a rapidly growing set of genomes. Right now, we have about 100,000, but the number is still growing exponentially.
We speak of the total number of <i>good genomes</i> as the universe.</p>

<p>The PheS protein is universal: there is normally one copy in each new genome. We are using a new measure of similarity: the
number of 8-mer amino acid sequences in common between the PheS proteins encoded in two genomes.
Let us say that 2 genomes are "similar" if they have 100 or more PheS 8-mers in common.
Then a representative set of genomes <b>G</b> is a subset of the universe that has two properties:</p>
<ol>
<li>Every genome in our universe is similar to at least one representative genome.</li>
<li>No two representative genomes are similar.</li>
</ol>
<p>You might find it irritating that a genome can be represented by multiple representative genomes. Do not let
this disturb you overmuch. What is important is that every genome in the universe has at least one representative,
no two representatives are similar, and that this implies that the universe is covered by representatives.</p>

<p>To generate this set, we start with a directory <b>Data.All</b> constructed as above with all good genomes in it.
That is, <b>Data.All/complete.genomes</b> is a list of the good genomes and <b>Data.All/6.1.1.20.fasta</b> is a FASTA file
of their PheS proteins. Then</p>
<pre class="call">
get_vss_for_reps Data.All
generate_cached_rep_data Data.All
</pre>
<p>builds <b>Data.All</b> into a fully functional <b>rep_server</b> directory.</p>
<pre class="call">
rep_server Data.All
?? rep_set 100 save=representative.genomes
?? x
</pre>
<p>Computes a representative genome set according to the criteria above. Now we want a <b>rep_server</b> directory for
this reduced set of genomes (the representatives).</p>
<pre class="call">
mkdir Data.Rep
cp representative.genomes Data.Rep/complete.genomes
cp Data.All/6.1.1.20.fasta Data.Rep
get_vss_for_reps Data.Rep
generate_cached_rep_data Data.Rep
</pre>
<p>The above creates a <b>rep_server</b> directory for just the representative genomes. Note that the FASTA file is too
big, but the software filters it on input, so this is not a problem.</p>

<p>We want to be able to examine a file of sequences and find out which representative sets are present. Use</p>
<pre class="call">
PheS_generate_signatures Data.Rep
</pre>

<p>to generate the file <b>Data.Rep/PheS.signatures</b> of protein signatures for the PheS proteins in the various representatives.
Once this is done, you can find the signatures in a DNA FASTA file by running it through <b>PheS_evidence_of_reps</b> as
follows</p>
<pre class="call">
PheS_evidence_of_reps Data.Rep &lt;input.fasta &gt;report.txt
</pre>

</div>
</body>
</html>